// @hubspot/api-client/lib/codegen/crm/objects/notes/types/ObjectParamAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSearchApi = exports.ObjectPublicObjectApi = exports.ObjectBatchApi = exports.ObjectBasicApi = exports.ObjectAssociationsApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class ObjectAssociationsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableAssociationsApi(configuration, requestFactory, responseProcessor);
    }
    archive(param, options) {
        return this.api.archive(param.noteId, param.toObjectType, param.toObjectId, param.associationType, options).toPromise();
    }
    create(param, options) {
        return this.api.create(param.noteId, param.toObjectType, param.toObjectId, param.associationType, options).toPromise();
    }
    getAll(param, options) {
        return this.api.getAll(param.noteId, param.toObjectType, param.after, param.limit, options).toPromise();
    }
}
exports.ObjectAssociationsApi = ObjectAssociationsApi;
const ObservableAPI_2 = require("./ObservableAPI");
class ObjectBasicApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_2.ObservableBasicApi(configuration, requestFactory, responseProcessor);
    }
    archive(param, options) {
        return this.api.archive(param.noteId, options).toPromise();
    }
    create(param, options) {
        return this.api.create(param.simplePublicObjectInput, options).toPromise();
    }
    getById(param, options) {
        return this.api.getById(param.noteId, param.properties, param.propertiesWithHistory, param.associations, param.archived, param.idProperty, options).toPromise();
    }
    getPage(param, options) {
        return this.api.getPage(param.limit, param.after, param.properties, param.propertiesWithHistory, param.associations, param.archived, options).toPromise();
    }
    update(param, options) {
        return this.api.update(param.noteId, param.simplePublicObjectInput, param.idProperty, options).toPromise();
    }
}
exports.ObjectBasicApi = ObjectBasicApi;
const ObservableAPI_3 = require("./ObservableAPI");
class ObjectBatchApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_3.ObservableBatchApi(configuration, requestFactory, responseProcessor);
    }
    archive(param, options) {
        return this.api.archive(param.batchInputSimplePublicObjectId, options).toPromise();
    }
    create(param, options) {
        return this.api.create(param.batchInputSimplePublicObjectInput, options).toPromise();
    }
    read(param, options) {
        return this.api.read(param.batchReadInputSimplePublicObjectId, param.archived, options).toPromise();
    }
    update(param, options) {
        return this.api.update(param.batchInputSimplePublicObjectBatchInput, options).toPromise();
    }
}
exports.ObjectBatchApi = ObjectBatchApi;
const ObservableAPI_4 = require("./ObservableAPI");
class ObjectPublicObjectApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_4.ObservablePublicObjectApi(configuration, requestFactory, responseProcessor);
    }
    merge(param, options) {
        return this.api.merge(param.publicMergeInput, options).toPromise();
    }
}
exports.ObjectPublicObjectApi = ObjectPublicObjectApi;
const ObservableAPI_5 = require("./ObservableAPI");
class ObjectSearchApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_5.ObservableSearchApi(configuration, requestFactory, responseProcessor);
    }
    doSearch(param, options) {
        return this.api.doSearch(param.publicObjectSearchRequest, options).toPromise();
    }
}
exports.ObjectSearchApi = ObjectSearchApi;
