// slack-block-builder/dist/internal/methods/set-methods.js
"use strict";
/* eslint-disable max-classes-per-file */
/* eslint-disable max-len */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Value = exports.Url = exports.Ts = exports.ThreadTs = exports.Title = exports.Text = exports.Submit = exports.PrivateMetaData = exports.PostAt = exports.Placeholder = exports.MinLength = exports.MinQueryLength = exports.MaxSelectedItems = exports.MaxLength = exports.Label = exports.InitialValue = exports.InitialUser = exports.InitialTime = exports.InitialOption = exports.InitialDate = exports.InitialConversation = exports.InitialChannel = exports.ImageUrl = exports.Hint = exports.Fallback = exports.ExternalId = exports.Element = exports.Description = exports.Deny = exports.Confirm = exports.Color = exports.Close = exports.Channel = exports.CallbackId = exports.BlockId = exports.AltText = exports.ActionId = exports.Accessory = exports.AccessibilityLabel = void 0;
const lib_1 = require("../lib");
const constants_1 = require("../constants");
class AccessibilityLabel extends lib_1.Builder {
    /**
     * @description A label for longer descriptive text about a button element. This label will be read out by screen readers instead of the button text object.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 75 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    accessibilityLabel(label) {
        return this.set(label, constants_1.Prop.AccessibilityLabel);
    }
}
exports.AccessibilityLabel = AccessibilityLabel;
class Accessory extends lib_1.Builder {
    /**
     * @description Adds an element to the section block of your view or message.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 1 element.
     *    * Can be any one of the elements.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    accessory(element) {
        return this.set(element, constants_1.Prop.Accessory);
    }
}
exports.Accessory = Accessory;
class ActionId extends lib_1.Builder {
    /**
     * @description Sets a string to be an identifier for the action taken by the user. It is sent back to your app in the interaction payload when the element is interacted or when the view is submitted.
     *
     * **Slack Validation Rules and Tips:**
     *    * **Required** ⚠
     *    * Each element in a view or message must have its own unique action ID.
     *    * Maximum of 255 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    actionId(actionId) {
        return this.set(actionId, constants_1.Prop.ActionId);
    }
}
exports.ActionId = ActionId;
class AltText extends lib_1.Builder {
    /**
     * @description This a plain-text summary of the image element or block.
     *
     * **Slack Validation Rules and Tips:**
     *    * **Required** ⚠
     *    * Should not contain any markup.
     *    * Maximum of 2000 characters.
     *
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    altText(altText) {
        return this.set(altText, constants_1.Prop.AltText);
    }
}
exports.AltText = AltText;
class BlockId extends lib_1.Builder {
    /**
     * @description Sets a string to be an identifier for any given block in a view or message. This is sent back to your app in interaction payloads and view submissions for your app to process.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 255 characters.
     *    * Each block in a view or message must have its own unique action ID.
     *    * If the contents of a block is updated, the block ID should also be updated.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    blockId(blockId) {
        return this.set(blockId, constants_1.Prop.BlockId);
    }
}
exports.BlockId = BlockId;
class CallbackId extends lib_1.Builder {
    /**
     * @description Sets a string for your view that is sent back to your server in all of the interaction payloads and view submissions. Used to identify the view from which the interaction payload is generated.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 255 characters.
     *    * It is recommended that sensitive data not be stored in the callback ID. Instead, use the `privateMetaData()` method.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    callbackId(callbackId) {
        return this.set(callbackId, constants_1.Prop.CallbackId);
    }
}
exports.CallbackId = CallbackId;
class Channel extends lib_1.Builder {
    /**
     * @description Sets the Slack channel ID to which the message will be sent via the API.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    channel(channel) {
        return this.set(channel, constants_1.Prop.Channel);
    }
}
exports.Channel = Channel;
class Close extends lib_1.Builder {
    /**
     * @description Sets the text displayed on the button that closes the view.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 24 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    close(closeButtonText) {
        return this.set(closeButtonText, constants_1.Prop.Close);
    }
}
exports.Close = Close;
class Color extends lib_1.Builder {
    /**
     * @description Sets the color for the blockquote border to the left of the attachment.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    color(color) {
        return this.set(color, constants_1.Prop.Color);
    }
}
exports.Color = Color;
class Confirm extends lib_1.Builder {
    /**
     * @description For confirmation dialogs, sets the text of the button that confirms the action to which the confirmation dialog has been added. For elements, adds a confirmation dialog that is displayed when the user interacts with the element to confirm the selection or action.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    confirm(value) {
        return this.set(value, constants_1.Prop.Confirm);
    }
}
exports.Confirm = Confirm;
class Deny extends lib_1.Builder {
    /**
     * @description Sets the text of the button that is meant to cancel the action or which the confirmation dialog was called.
     *
     * **Slack Validation Rules and Tips:**
     *    * **Required** ⚠
     *    * Maximum of 30 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    deny(text) {
        return this.set(text, constants_1.Prop.Deny);
    }
}
exports.Deny = Deny;
class Description extends lib_1.Builder {
    /**
     * @description Sets the descriptive text displayed below the text field of the option.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 75 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    description(description) {
        return this.set(description, constants_1.Prop.Description);
    }
}
exports.Description = Description;
class Element extends lib_1.Builder {
    /**
     * @description Adds a user input element to the input block for gathering information from the user.
     *
     * **Slack Validation Rules and Tips:**
     *    * **Required** ⚠
     *    * Maximum of 1 element.
     *    * Supports text inputs, select and multi-select menus, as well as date pickers and checkbox inputs.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    element(element) {
        return this.set(element, constants_1.Prop.Element);
    }
}
exports.Element = Element;
class ExternalId extends lib_1.Builder {
    /**
     * @description Sets a custom identifier for a view or file that must be unique for all views on a per-team basis.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 255 characters.
     *    * When used, an external ID must be unique to a certain view.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    externalId(externalId) {
        return this.set(externalId, constants_1.Prop.ExternalId);
    }
}
exports.ExternalId = ExternalId;
class Fallback extends lib_1.Builder {
    /**
     * @description Sets the plain text summary of the attachment used in clients that can't display formatted text (eg. IRC, mobile notifications).
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    fallback(text) {
        return this.set(text, constants_1.Prop.Fallback);
    }
}
exports.Fallback = Fallback;
class Hint extends lib_1.Builder {
    /**
     * @description Adds a hint below the input block to provide users with more context or instructions for using the input.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 2000 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    hint(hint) {
        return this.set(hint, constants_1.Prop.Hint);
    }
}
exports.Hint = Hint;
class ImageUrl extends lib_1.Builder {
    /**
     * @description Sets the source URL for the image block or element that you wish to include in your view or message.
     *
     * **Slack Validation Rules and Tips:**
     *    * **Required** ⚠
     *    * Maximum of 2000 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    imageUrl(url) {
        return this.set(url, constants_1.Prop.ImageUrl);
    }
}
exports.ImageUrl = ImageUrl;
class InitialChannel extends lib_1.Builder {
    /**
     * @description Pre-populates the menu with a selected, default channel.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    initialChannel(channelId) {
        return this.set(channelId, constants_1.Prop.InitialChannel);
    }
}
exports.InitialChannel = InitialChannel;
class InitialConversation extends lib_1.Builder {
    /**
     * @description Pre-populates the menu with a selected, default conversation.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    initialConversation(conversationId) {
        return this.set(conversationId, constants_1.Prop.InitialConversation);
    }
}
exports.InitialConversation = InitialConversation;
class InitialDate extends lib_1.Builder {
    /**
     * @description Pre-populates the date picker with a selected, default date.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    initialDate(date) {
        return this.set(date, constants_1.Prop.InitialDate);
    }
}
exports.InitialDate = InitialDate;
class InitialOption extends lib_1.Builder {
    /**
     * @description Pre-populates the menu or date picker with a selected, default option.
     *
     * **Slack Validation Rules and Tips:**
     *    * Must be an exact match for one of the options in the menu.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    initialOption(option) {
        return this.set(option, constants_1.Prop.InitialOption);
    }
}
exports.InitialOption = InitialOption;
class InitialTime extends lib_1.Builder {
    /**
     * @description Pre-populates the time picker with a selected, default time.
     *
     * **Slack Validation Rules and Tips:**
     *    * Set in HH:mm format, where HH is 24-hour hour format and mm is minutes with a leading zero.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    initialTime(time) {
        return this.set(time, constants_1.Prop.InitialTime);
    }
}
exports.InitialTime = InitialTime;
class InitialUser extends lib_1.Builder {
    /**
     * @description Pre-populates the menu with a selected, default user.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    initialUser(userId) {
        return this.set(userId, constants_1.Prop.InitialUser);
    }
}
exports.InitialUser = InitialUser;
class InitialValue extends lib_1.Builder {
    /**
     * @description Pre-populates the text input with default, editable text.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    initialValue(value) {
        return this.set(value, constants_1.Prop.InitialValue);
    }
}
exports.InitialValue = InitialValue;
class Label extends lib_1.Builder {
    /**
     * @description For input blocks, this sets the text displayed in the label for the input. For an option group, it defines the text displayed as a title for the group.
     *
     * **Slack Validation Rules and Tips:**
     *    * **Required** ⚠
     *    * For input blocks, maximum of 2000 characters.
     *    * For option groups, maximum of 75 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    label(label) {
        return this.set(label, constants_1.Prop.Label);
    }
}
exports.Label = Label;
class MaxLength extends lib_1.Builder {
    /**
     * @description Sets a maximum character count allowed in the given text input.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    maxLength(length) {
        return this.set(length, constants_1.Prop.MaxLength);
    }
}
exports.MaxLength = MaxLength;
class MaxSelectedItems extends lib_1.Builder {
    /**
     * @description Sets a limit to how many items the user can select in the multi-select menu.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    maxSelectedItems(quantity) {
        return this.set(quantity, constants_1.Prop.MaxSelectedItems);
    }
}
exports.MaxSelectedItems = MaxSelectedItems;
class MinQueryLength extends lib_1.Builder {
    /**
     * @description Sets a minimum number of characters required before Slack queries your app for a list of options.
     *
     * **Slack Validation Rules and Tips:**
     *    * If not set, the request will be sent on every character entered or removed.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    minQueryLength(length) {
        return this.set(length, constants_1.Prop.MinQueryLength);
    }
}
exports.MinQueryLength = MinQueryLength;
class MinLength extends lib_1.Builder {
    /**
     * @description Sets a minimum character count required for the given text input before the user can submit the view.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum 3000 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    minLength(length) {
        return this.set(length, constants_1.Prop.MinLength);
    }
}
exports.MinLength = MinLength;
class Placeholder extends lib_1.Builder {
    /**
     * @description Defines the text displayed as a placeholder in the empty input element.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 150 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    placeholder(placeholder) {
        return this.set(placeholder, constants_1.Prop.Placeholder);
    }
}
exports.Placeholder = Placeholder;
class PostAt extends lib_1.Builder {
    /**
     * @description Sets a time in the future for the message to be sent to the channel or user, as a scheduled message.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    postAt(timestamp) {
        return this.set(timestamp, constants_1.Prop.PostAt);
    }
}
exports.PostAt = PostAt;
class PrivateMetaData extends lib_1.Builder {
    /**
     * @description Defines a string sent back to your server with view and interaction payloads.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum 3000 characters.
     *    * Typically used to persist data or store context between views.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    privateMetaData(privateMetaData) {
        return this.set(privateMetaData, constants_1.Prop.PrivateMetaData);
    }
}
exports.PrivateMetaData = PrivateMetaData;
class Submit extends lib_1.Builder {
    /**
     * @description Sets the text displayed on the button that submits the view.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 24 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    submit(text) {
        return this.set(text, constants_1.Prop.Submit);
    }
}
exports.Submit = Submit;
class Text extends lib_1.Builder {
    /**
     * @description Sets the text displayed for buttons, headers, confirmation dialogs, sections, context blocks, and options.
     *
     * **Slack Validation Rules and Tips:**
     *    * **Required for buttons, headers, confirmation dialogs, and option** ⚠
     *    * **Required for sections if fields are not defined** ⚠
     *    * For buttons and options, maximum of 75 characters.
     *    * For confirmation dialogs, maximum of 75 characters.
     *    * For section and header blocks, maximum of 3000 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    text(text) {
        return this.set(text, constants_1.Prop.Text);
    }
}
exports.Text = Text;
class Title extends lib_1.Builder {
    /**
     * @description Sets the title displayed for the block, element, or confirmation dialog.
     *
     * **Slack Validation Rules and Tips:**
     *    * **Required for views and confirmation dialogs** ⚠
     *    * For views, maximum of 24 characters.
     *    * For images, maximum of 2000 characters.
     *    * For confirmation dialogs, maximum of 100 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    title(title) {
        return this.set(title, constants_1.Prop.Title);
    }
}
exports.Title = Title;
class ThreadTs extends lib_1.Builder {
    /**
     * @description Instructs the Slack API to send the message to the thread of the message associated with the timestamp.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    threadTs(threadTs) {
        return this.set(threadTs, constants_1.Prop.ThreadTs);
    }
}
exports.ThreadTs = ThreadTs;
class Ts extends lib_1.Builder {
    /**
     * @description Instructs the Slack API to use the message to replaced an existing message.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    ts(ts) {
        return this.set(ts, constants_1.Prop.Ts);
    }
}
exports.Ts = Ts;
class Url extends lib_1.Builder {
    /**
     * @description Sets the URL to which the user is redirected when interacting with a button or option.
     *
     * **Slack Validation Rules and Tips:**
     *    * Maximum of 3000 characters.
     *    * For options, it is only supported for options in an overflow menu.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    url(url) {
        return this.set(url, constants_1.Prop.Url);
    }
}
exports.Url = Url;
class Value extends lib_1.Builder {
    /**
     * @description Sets a value to be sent to your app when a user interacts with a button or option.
     *
     * **Slack Validation Rules and Tips:**
     *    * For buttons, maximum of 2000 characters.
     *    * For options, maximum of 75 characters.
     *
     * {@link https://api.slack.com/block-kit|Open Official Slack Block Kit Documentation}
     * {@link https://www.blockbuilder.dev|Open Block Builder Documentation}
     */
    value(value) {
        return this.set(value, constants_1.Prop.Value);
    }
}
exports.Value = Value;
