// @hubspot/api-client/lib/codegen/crm/deals/auth/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configureAuthMethods = exports.Oauth2Authentication = exports.HapikeyAuthentication = void 0;
class HapikeyAuthentication {
    constructor(apiKey) {
        this.apiKey = apiKey;
    }
    getName() {
        return "hapikey";
    }
    applySecurityAuthentication(context) {
        context.setQueryParam("hapikey", this.apiKey);
    }
}
exports.HapikeyAuthentication = HapikeyAuthentication;
class Oauth2Authentication {
    constructor(accessToken) {
        this.accessToken = accessToken;
    }
    getName() {
        return "oauth2";
    }
    applySecurityAuthentication(context) {
        context.setHeaderParam("Authorization", "Bearer " + this.accessToken);
    }
}
exports.Oauth2Authentication = Oauth2Authentication;
function configureAuthMethods(config) {
    let authMethods = {};
    if (!config) {
        return authMethods;
    }
    if (config["hapikey"]) {
        authMethods["hapikey"] = new HapikeyAuthentication(config["hapikey"]);
    }
    if (config["oauth2"]) {
        authMethods["oauth2"] = new Oauth2Authentication(config["oauth2"]["accessToken"]);
    }
    return authMethods;
}
exports.configureAuthMethods = configureAuthMethods;
