// @hubspot/api-client/lib/src/discovery/crm/objects/calls/CallsDiscovery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallsDiscovery = void 0;
const configuration_1 = require("../../../../../codegen/crm/objects/calls/configuration");
const index_1 = require("../../../../../codegen/crm/objects/calls/index");
const rxjsStub_1 = require("../../../../../codegen/crm/objects/calls/rxjsStub");
const ApiClientConfigurator_1 = require("../../../../configuration/ApiClientConfigurator");
class CallsDiscovery {
    constructor(config) {
        const configuration = configuration_1.createConfiguration(ApiClientConfigurator_1.ApiClientConfigurator.getParams(config, index_1.ServerConfiguration, rxjsStub_1.Observable, rxjsStub_1.Observable));
        this.associationsApi = new index_1.AssociationsApi(configuration);
        this.basicApi = new index_1.BasicApi(configuration);
        this.batchApi = new index_1.BatchApi(configuration);
        this.publicObjectApi = new index_1.PublicObjectApi(configuration);
        this.searchApi = new index_1.SearchApi(configuration);
    }
}
exports.CallsDiscovery = CallsDiscovery;
