// slack-block-builder/dist/internal/constants/element-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElementType = void 0;
var ElementType;
(function (ElementType) {
    ElementType["Button"] = "button";
    ElementType["Checkboxes"] = "checkboxes";
    ElementType["DatePicker"] = "datepicker";
    ElementType["TimePicker"] = "timepicker";
    ElementType["Image"] = "image";
    ElementType["Overflow"] = "overflow";
    ElementType["TextInput"] = "plain_text_input";
    ElementType["RadioButtons"] = "radio_buttons";
    ElementType["StaticSelect"] = "static_select";
    ElementType["ExternalSelect"] = "external_select";
    ElementType["UserSelect"] = "users_select";
    ElementType["ConversationSelect"] = "conversations_select";
    ElementType["ChannelSelect"] = "channels_select";
    ElementType["StaticMultiSelect"] = "multi_static_select";
    ElementType["ExternalMultiSelect"] = "multi_external_select";
    ElementType["UserMultiSelect"] = "multi_users_select";
    ElementType["ConversationsMultiSelect"] = "multi_conversations_select";
    ElementType["ChannelsMultiSelect"] = "multi_channels_select";
})(ElementType = exports.ElementType || (exports.ElementType = {}));
