// slack-block-builder/dist/internal/constants/props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Prop = void 0;
var Prop;
(function (Prop) {
    Prop["Blocks"] = "blocks";
    Prop["Elements"] = "elements";
    Prop["BlockId"] = "blockId";
    Prop["ExternalId"] = "externalId";
    Prop["Label"] = "label";
    Prop["Element"] = "element";
    Prop["Hint"] = "hint";
    Prop["Optional"] = "optional";
    Prop["Fields"] = "fields";
    Prop["Accessory"] = "accessory";
    Prop["ActionId"] = "actionId";
    Prop["Url"] = "url";
    Prop["Style"] = "style";
    Prop["Value"] = "value";
    Prop["Option"] = "option";
    Prop["Confirm"] = "confirm";
    Prop["ImageUrl"] = "imageUrl";
    Prop["AltText"] = "altText";
    Prop["Options"] = "options";
    Prop["InitialOptions"] = "initialOptions";
    Prop["InitialOption"] = "initialOption";
    Prop["Placeholder"] = "placeholder";
    Prop["InitialDate"] = "initialDate";
    Prop["InitialValue"] = "initialValue";
    Prop["Multiline"] = "multiline";
    Prop["MinLength"] = "minLength";
    Prop["MaxLength"] = "maxLength";
    Prop["InitialChannel"] = "initialChannel";
    Prop["InitialChannels"] = "initialChannels";
    Prop["InitialConversation"] = "initialConversation";
    Prop["InitialConversations"] = "initialConversations";
    Prop["ResponseUrlEnabled"] = "responseUrlEnabled";
    Prop["DefaultToCurrentConversation"] = "defaultToCurrentConversation";
    Prop["Filter"] = "filter";
    Prop["MinQueryLength"] = "minQueryLength";
    Prop["OptionGroups"] = "optionGroups";
    Prop["InitialUser"] = "initialUser";
    Prop["InitialUsers"] = "initialUsers";
    Prop["MaxSelectedItems"] = "maxSelectedItems";
    Prop["Title"] = "title";
    Prop["Submit"] = "submit";
    Prop["Close"] = "close";
    Prop["Deny"] = "deny";
    Prop["ExcludeExternalSharedChannels"] = "excludeExternalSharedChannels";
    Prop["ExcludeBotUsers"] = "excludeBotUsers";
    Prop["Text"] = "text";
    Prop["PrivateMetaData"] = "privateMetaData";
    Prop["CallbackId"] = "callbackId";
    Prop["Channel"] = "channel";
    Prop["ClearOnClose"] = "clearOnClose";
    Prop["NotifyOnClose"] = "notifyOnClose";
    Prop["Description"] = "description";
    Prop["Danger"] = "danger";
    Prop["Primary"] = "primary";
    Prop["AsUser"] = "asUser";
    Prop["ThreadTs"] = "threadTs";
    Prop["ReplaceOriginal"] = "replaceOriginal";
    Prop["DeleteOriginal"] = "deleteOriginal";
    Prop["ResponseType"] = "responseType";
    Prop["PostAt"] = "postAt";
    Prop["Ephemeral"] = "ephemeral";
    Prop["InChannel"] = "inChannel";
    Prop["Ts"] = "ts";
    Prop["Color"] = "color";
    Prop["Fallback"] = "fallback";
    Prop["Attachments"] = "attachments";
    Prop["DispatchAction"] = "dispatchAction";
    Prop["DispatchActionConfig"] = "dispatchActionConfig";
    Prop["OnEnterPressed"] = "onEnterPressed";
    Prop["OnCharacterEntered"] = "onCharacterEntered";
    Prop["DispatchActionOnEnterPressed"] = "dispatchActionOnEnterPressed";
    Prop["DispatchActionOnCharacterEntered"] = "dispatchActionOnCharacterEntered";
    Prop["InitialTime"] = "initialTime";
    Prop["Mrkdwn"] = "mrkdwn";
    Prop["IgnoreMarkdown"] = "ignoreMarkdown";
    Prop["SubmitDisabled"] = "submitDisabled";
    Prop["FocusOnLoad"] = "focusOnLoad";
    Prop["AccessibilityLabel"] = "accessibilityLabel";
})(Prop = exports.Prop || (exports.Prop = {}));
