// slack-block-builder/dist/bits/confirmation-dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfirmationDialogBuilder = void 0;
const base_1 = require("../internal/base");
const dto_1 = require("../internal/dto");
const helpers_1 = require("../internal/helpers");
const methods_1 = require("../internal/methods");
/**
 * @@link https://api.slack.com/reference/block-kit/composition-objects#confirm
 * @@displayName Confirmation Dialog
 */
class ConfirmationDialogBuilder extends base_1.BitBuilderBase {
    /** @internal */
    build() {
        return this.getResult(dto_1.SlackDto, {
            text: helpers_1.getMarkdownObject(this.props.text),
            title: helpers_1.getPlainTextObject(this.props.title),
            confirm: helpers_1.getPlainTextObject(this.props.confirm),
            deny: helpers_1.getPlainTextObject(this.props.deny),
        });
    }
}
exports.ConfirmationDialogBuilder = ConfirmationDialogBuilder;
helpers_1.applyMixins(ConfirmationDialogBuilder, [
    methods_1.Confirm,
    methods_1.Danger,
    methods_1.Deny,
    methods_1.End,
    methods_1.Primary,
    methods_1.Text,
    methods_1.Title,
]);
