// @tryvital/vital-node/dist/client/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevicesAPI = exports.ProfileApi = exports.TestkitsApi = exports.WorkoutsApi = exports.WebhooksApi = exports.UserApi = exports.SleepApi = exports.LinkApi = exports.BodyApi = exports.ActivityApi = void 0;
var Activity_1 = require("./Activity");
Object.defineProperty(exports, "ActivityApi", { enumerable: true, get: function () { return Activity_1.ActivityApi; } });
var Body_1 = require("./Body");
Object.defineProperty(exports, "BodyApi", { enumerable: true, get: function () { return Body_1.BodyApi; } });
var Link_1 = require("./Link");
Object.defineProperty(exports, "LinkApi", { enumerable: true, get: function () { return Link_1.LinkApi; } });
var Sleep_1 = require("./Sleep");
Object.defineProperty(exports, "SleepApi", { enumerable: true, get: function () { return Sleep_1.SleepApi; } });
var User_1 = require("./User");
Object.defineProperty(exports, "UserApi", { enumerable: true, get: function () { return User_1.UserApi; } });
var Webhooks_1 = require("./Webhooks");
Object.defineProperty(exports, "WebhooksApi", { enumerable: true, get: function () { return Webhooks_1.WebhooksApi; } });
var Workouts_1 = require("./Workouts");
Object.defineProperty(exports, "WorkoutsApi", { enumerable: true, get: function () { return Workouts_1.WorkoutsApi; } });
var Testkits_1 = require("./Testkits");
Object.defineProperty(exports, "TestkitsApi", { enumerable: true, get: function () { return Testkits_1.TestkitsApi; } });
var Profile_1 = require("./Profile");
Object.defineProperty(exports, "ProfileApi", { enumerable: true, get: function () { return Profile_1.ProfileApi; } });
var Devices_1 = require("./Devices");
Object.defineProperty(exports, "DevicesAPI", { enumerable: true, get: function () { return Devices_1.DevicesAPI; } });
