// @hubspot/api-client/lib/codegen/automation/actions/models/ObjectSerializer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSerializer = void 0;
__exportStar(require("./ActionFunction"), exports);
__exportStar(require("./ActionFunctionIdentifier"), exports);
__exportStar(require("./ActionLabels"), exports);
__exportStar(require("./ActionRevision"), exports);
__exportStar(require("./BatchInputCallbackCompletionBatchRequest"), exports);
__exportStar(require("./CallbackCompletionBatchRequest"), exports);
__exportStar(require("./CallbackCompletionRequest"), exports);
__exportStar(require("./CollectionResponseActionFunctionIdentifierNoPaging"), exports);
__exportStar(require("./CollectionResponseActionRevisionForwardPaging"), exports);
__exportStar(require("./CollectionResponseExtensionActionDefinitionForwardPaging"), exports);
__exportStar(require("./ConditionalSingleFieldDependency"), exports);
__exportStar(require("./ErrorDetail"), exports);
__exportStar(require("./ExtensionActionDefinition"), exports);
__exportStar(require("./ExtensionActionDefinitionInput"), exports);
__exportStar(require("./ExtensionActionDefinitionPatch"), exports);
__exportStar(require("./FieldTypeDefinition"), exports);
__exportStar(require("./ForwardPaging"), exports);
__exportStar(require("./InputFieldDefinition"), exports);
__exportStar(require("./ModelError"), exports);
__exportStar(require("./NextPage"), exports);
__exportStar(require("./ObjectRequestOptions"), exports);
__exportStar(require("./Option"), exports);
__exportStar(require("./SingleFieldDependency"), exports);
const ActionFunction_1 = require("./ActionFunction");
const ActionFunctionIdentifier_1 = require("./ActionFunctionIdentifier");
const ActionLabels_1 = require("./ActionLabels");
const ActionRevision_1 = require("./ActionRevision");
const BatchInputCallbackCompletionBatchRequest_1 = require("./BatchInputCallbackCompletionBatchRequest");
const CallbackCompletionBatchRequest_1 = require("./CallbackCompletionBatchRequest");
const CallbackCompletionRequest_1 = require("./CallbackCompletionRequest");
const CollectionResponseActionFunctionIdentifierNoPaging_1 = require("./CollectionResponseActionFunctionIdentifierNoPaging");
const CollectionResponseActionRevisionForwardPaging_1 = require("./CollectionResponseActionRevisionForwardPaging");
const CollectionResponseExtensionActionDefinitionForwardPaging_1 = require("./CollectionResponseExtensionActionDefinitionForwardPaging");
const ConditionalSingleFieldDependency_1 = require("./ConditionalSingleFieldDependency");
const ErrorDetail_1 = require("./ErrorDetail");
const ExtensionActionDefinition_1 = require("./ExtensionActionDefinition");
const ExtensionActionDefinitionInput_1 = require("./ExtensionActionDefinitionInput");
const ExtensionActionDefinitionPatch_1 = require("./ExtensionActionDefinitionPatch");
const FieldTypeDefinition_1 = require("./FieldTypeDefinition");
const ForwardPaging_1 = require("./ForwardPaging");
const InputFieldDefinition_1 = require("./InputFieldDefinition");
const ModelError_1 = require("./ModelError");
const NextPage_1 = require("./NextPage");
const ObjectRequestOptions_1 = require("./ObjectRequestOptions");
const Option_1 = require("./Option");
const SingleFieldDependency_1 = require("./SingleFieldDependency");
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
const supportedMediaTypes = {
    "application/json": Infinity,
    "application/octet-stream": 0,
    "application/x-www-form-urlencoded": 0
};
let enumsMap = new Set([
    "ActionFunctionFunctionTypeEnum",
    "ActionFunctionIdentifierFunctionTypeEnum",
    "ConditionalSingleFieldDependencyDependencyTypeEnum",
    "FieldTypeDefinitionTypeEnum",
    "FieldTypeDefinitionFieldTypeEnum",
    "FieldTypeDefinitionReferencedObjectTypeEnum",
    "InputFieldDefinitionSupportedValueTypesEnum",
    "SingleFieldDependencyDependencyTypeEnum",
]);
let typeMap = {
    "ActionFunction": ActionFunction_1.ActionFunction,
    "ActionFunctionIdentifier": ActionFunctionIdentifier_1.ActionFunctionIdentifier,
    "ActionLabels": ActionLabels_1.ActionLabels,
    "ActionRevision": ActionRevision_1.ActionRevision,
    "BatchInputCallbackCompletionBatchRequest": BatchInputCallbackCompletionBatchRequest_1.BatchInputCallbackCompletionBatchRequest,
    "CallbackCompletionBatchRequest": CallbackCompletionBatchRequest_1.CallbackCompletionBatchRequest,
    "CallbackCompletionRequest": CallbackCompletionRequest_1.CallbackCompletionRequest,
    "CollectionResponseActionFunctionIdentifierNoPaging": CollectionResponseActionFunctionIdentifierNoPaging_1.CollectionResponseActionFunctionIdentifierNoPaging,
    "CollectionResponseActionRevisionForwardPaging": CollectionResponseActionRevisionForwardPaging_1.CollectionResponseActionRevisionForwardPaging,
    "CollectionResponseExtensionActionDefinitionForwardPaging": CollectionResponseExtensionActionDefinitionForwardPaging_1.CollectionResponseExtensionActionDefinitionForwardPaging,
    "ConditionalSingleFieldDependency": ConditionalSingleFieldDependency_1.ConditionalSingleFieldDependency,
    "ErrorDetail": ErrorDetail_1.ErrorDetail,
    "ExtensionActionDefinition": ExtensionActionDefinition_1.ExtensionActionDefinition,
    "ExtensionActionDefinitionInput": ExtensionActionDefinitionInput_1.ExtensionActionDefinitionInput,
    "ExtensionActionDefinitionPatch": ExtensionActionDefinitionPatch_1.ExtensionActionDefinitionPatch,
    "FieldTypeDefinition": FieldTypeDefinition_1.FieldTypeDefinition,
    "ForwardPaging": ForwardPaging_1.ForwardPaging,
    "InputFieldDefinition": InputFieldDefinition_1.InputFieldDefinition,
    "ModelError": ModelError_1.ModelError,
    "NextPage": NextPage_1.NextPage,
    "ObjectRequestOptions": ObjectRequestOptions_1.ObjectRequestOptions,
    "Option": Option_1.Option,
    "SingleFieldDependency": SingleFieldDependency_1.SingleFieldDependency,
};
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap.has(expectedType)) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType;
            }
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType;
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    if (typeMap[discriminatorType]) {
                        return discriminatorType;
                    }
                    else {
                        return expectedType;
                    }
                }
                else {
                    return expectedType;
                }
            }
        }
    }
    static serialize(data, type, format) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) {
            let subType = type.replace("Array<", "");
            subType = subType.substring(0, subType.length - 1);
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.serialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type === "Date") {
            if (format == "date") {
                let month = data.getMonth() + 1;
                month = month < 10 ? "0" + month.toString() : month.toString();
                let day = data.getDate();
                day = day < 10 ? "0" + day.toString() : day.toString();
                return data.getFullYear() + "-" + month + "-" + day;
            }
            else {
                return data.toISOString();
            }
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) {
                return data;
            }
            type = this.findCorrectType(data, type);
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static deserialize(data, type, format) {
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) {
            let subType = type.replace("Array<", "");
            subType = subType.substring(0, subType.length - 1);
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.deserialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) {
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.name] = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static normalizeMediaType(mediaType) {
        if (mediaType === undefined) {
            return undefined;
        }
        return mediaType.split(";")[0].trim().toLowerCase();
    }
    static getPreferredMediaType(mediaTypes) {
        if (!mediaTypes) {
            return "application/json";
        }
        const normalMediaTypes = mediaTypes.map(this.normalizeMediaType);
        let selectedMediaType = undefined;
        let selectedRank = -Infinity;
        for (const mediaType of normalMediaTypes) {
            if (supportedMediaTypes[mediaType] > selectedRank) {
                selectedMediaType = mediaType;
                selectedRank = supportedMediaTypes[mediaType];
            }
        }
        if (selectedMediaType === undefined) {
            throw new Error("None of the given media types are supported: " + mediaTypes.join(", "));
        }
        return selectedMediaType;
    }
    static stringify(data, mediaType) {
        if (mediaType === "application/json") {
            return JSON.stringify(data);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.stringify.");
    }
    static parse(rawData, mediaType) {
        if (mediaType === undefined) {
            throw new Error("Cannot parse content. No Content-Type defined.");
        }
        if (mediaType === "application/json") {
            return JSON.parse(rawData);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.parse.");
    }
}
exports.ObjectSerializer = ObjectSerializer;
