// slack-block-builder/dist/elements/checkboxes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckboxesBuilder = void 0;
const base_1 = require("../internal/base");
const constants_1 = require("../internal/constants");
const dto_1 = require("../internal/dto");
const helpers_1 = require("../internal/helpers");
const methods_1 = require("../internal/methods");
/**
 * @@link https://api.slack.com/reference/block-kit/block-elements#checkboxes
 * @@displayName Checkboxes
 */
class CheckboxesBuilder extends base_1.ElementBuilderBase {
    /** @internal */
    build() {
        return this.getResult(dto_1.SlackElementDto, {
            type: constants_1.ElementType.Checkboxes,
            options: helpers_1.getBuilderResults(this.props.options, { isMarkdown: true }),
            initialOptions: helpers_1.getBuilderResults(this.props.initialOptions, { isMarkdown: true }),
            confirm: helpers_1.getBuilderResult(this.props.confirm),
        });
    }
}
exports.CheckboxesBuilder = CheckboxesBuilder;
helpers_1.applyMixins(CheckboxesBuilder, [
    methods_1.ActionId,
    methods_1.Confirm,
    methods_1.End,
    methods_1.FocusOnLoad,
    methods_1.InitialOptions,
    methods_1.Options,
]);
