// slack-block-builder/dist/elements/conversation-multi-select.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConversationMultiSelectBuilder = void 0;
const base_1 = require("../internal/base");
const constants_1 = require("../internal/constants");
const dto_1 = require("../internal/dto");
const helpers_1 = require("../internal/helpers");
const methods_1 = require("../internal/methods");
/**
 * @@link https://api.slack.com/reference/block-kit/block-elements#conversation_multi_select
 * @@displayName Conversation Multi-Select
 */
class ConversationMultiSelectBuilder extends base_1.ElementBuilderBase {
    /** @internal */
    build() {
        return this.getResult(dto_1.SlackElementDto, {
            type: constants_1.ElementType.ConversationsMultiSelect,
            placeholder: helpers_1.getPlainTextObject(this.props.placeholder),
            confirm: helpers_1.getBuilderResult(this.props.confirm),
            filter: helpers_1.getFilter(this.props),
        });
    }
}
exports.ConversationMultiSelectBuilder = ConversationMultiSelectBuilder;
helpers_1.applyMixins(ConversationMultiSelectBuilder, [
    methods_1.ActionId,
    methods_1.Confirm,
    methods_1.DefaultToCurrentConversation,
    methods_1.End,
    methods_1.ExcludeBotUsers,
    methods_1.ExcludeExternalSharedChannels,
    methods_1.Filter,
    methods_1.FocusOnLoad,
    methods_1.InitialConversations,
    methods_1.MaxSelectedItems,
    methods_1.Placeholder,
]);
