// slack-block-builder/dist/elements/datepicker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatePickerBuilder = void 0;
const base_1 = require("../internal/base");
const constants_1 = require("../internal/constants");
const dto_1 = require("../internal/dto");
const helpers_1 = require("../internal/helpers");
const methods_1 = require("../internal/methods");
/**
 * @@link https://api.slack.com/reference/block-kit/block-elements#datepicker
 * @@displayName Date Picker
 */
class DatePickerBuilder extends base_1.ElementBuilderBase {
    /** @internal */
    build() {
        return this.getResult(dto_1.SlackElementDto, {
            type: constants_1.ElementType.DatePicker,
            placeholder: helpers_1.getPlainTextObject(this.props.placeholder),
            initialDate: helpers_1.getFormattedDate(this.props.initialDate),
            confirm: helpers_1.getBuilderResult(this.props.confirm),
        });
    }
}
exports.DatePickerBuilder = DatePickerBuilder;
helpers_1.applyMixins(DatePickerBuilder, [
    methods_1.ActionId,
    methods_1.Confirm,
    methods_1.End,
    methods_1.FocusOnLoad,
    methods_1.InitialDate,
    methods_1.Placeholder,
]);
