// slack-block-builder/dist/elements/external-select.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalSelectBuilder = void 0;
const base_1 = require("../internal/base");
const constants_1 = require("../internal/constants");
const dto_1 = require("../internal/dto");
const helpers_1 = require("../internal/helpers");
const methods_1 = require("../internal/methods");
/**
 * @@link https://api.slack.com/reference/block-kit/block-elements#external_select
 * @@displayName External Select
 */
class ExternalSelectBuilder extends base_1.ElementBuilderBase {
    /** @internal */
    build() {
        return this.getResult(dto_1.SlackElementDto, {
            type: constants_1.ElementType.ExternalSelect,
            placeholder: helpers_1.getPlainTextObject(this.props.placeholder),
            initialOption: helpers_1.getBuilderResult(this.props.initialOption),
            confirm: helpers_1.getBuilderResult(this.props.confirm),
        });
    }
}
exports.ExternalSelectBuilder = ExternalSelectBuilder;
helpers_1.applyMixins(ExternalSelectBuilder, [
    methods_1.ActionId,
    methods_1.Confirm,
    methods_1.End,
    methods_1.FocusOnLoad,
    methods_1.InitialOption,
    methods_1.MinQueryLength,
    methods_1.Placeholder,
]);
