// slack-block-builder/dist/elements/static-select.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticSelectBuilder = void 0;
const base_1 = require("../internal/base");
const constants_1 = require("../internal/constants");
const dto_1 = require("../internal/dto");
const helpers_1 = require("../internal/helpers");
const methods_1 = require("../internal/methods");
/**
 * @@link https://api.slack.com/reference/block-kit/block-elements#static_select
 * @@displayName Static Select
 */
class StaticSelectBuilder extends base_1.ElementBuilderBase {
    /** @internal */
    build() {
        return this.getResult(dto_1.SlackElementDto, {
            type: constants_1.ElementType.StaticSelect,
            placeholder: helpers_1.getPlainTextObject(this.props.placeholder),
            options: helpers_1.getBuilderResults(this.props.options),
            optionGroups: helpers_1.getBuilderResults(this.props.optionGroups),
            initialOption: helpers_1.getBuilderResult(this.props.initialOption),
            confirm: helpers_1.getBuilderResult(this.props.confirm),
        });
    }
}
exports.StaticSelectBuilder = StaticSelectBuilder;
helpers_1.applyMixins(StaticSelectBuilder, [
    methods_1.ActionId,
    methods_1.Confirm,
    methods_1.End,
    methods_1.FocusOnLoad,
    methods_1.InitialOption,
    methods_1.OptionGroups,
    methods_1.Options,
    methods_1.Placeholder,
]);
