// slack-block-builder/dist/elements/static-multi-select.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticMultiSelectBuilder = void 0;
const base_1 = require("../internal/base");
const constants_1 = require("../internal/constants");
const dto_1 = require("../internal/dto");
const helpers_1 = require("../internal/helpers");
const methods_1 = require("../internal/methods");
class StaticMultiSelectBuilder extends base_1.ElementBuilderBase {
    /** @internal */
    build() {
        return this.getResult(dto_1.SlackElementDto, {
            type: constants_1.ElementType.StaticMultiSelect,
            placeholder: helpers_1.getPlainTextObject(this.props.placeholder),
            options: helpers_1.getBuilderResults(this.props.options),
            initialOptions: helpers_1.getBuilderResults(this.props.initialOptions),
            optionGroups: helpers_1.getBuilderResults(this.props.optionGroups),
            confirm: helpers_1.getBuilderResult(this.props.confirm),
        });
    }
}
exports.StaticMultiSelectBuilder = StaticMultiSelectBuilder;
helpers_1.applyMixins(StaticMultiSelectBuilder, [
    methods_1.ActionId,
    methods_1.Confirm,
    methods_1.End,
    methods_1.FocusOnLoad,
    methods_1.InitialOptions,
    methods_1.MaxSelectedItems,
    methods_1.OptionGroups,
    methods_1.Options,
    methods_1.Placeholder,
]);
