// @snaplet/copycat/dist/primitives.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var primitives_exports = {};
__export(primitives_exports, {
  bool: () => bool,
  char: () => char,
  dateString: () => dateString,
  digit: () => digit,
  float: () => float,
  hex: () => hex,
  int: () => int,
  oneOf: () => oneOf,
  paragraph: () => paragraph,
  sentence: () => sentence,
  times: () => times,
  word: () => word,
  words: () => words
});
module.exports = __toCommonJS(primitives_exports);
var import_fictional = __toESM(require("fictional"));
const int = import_fictional.default.int;
const bool = import_fictional.default.bool;
const float = import_fictional.default.float;
const dateString = import_fictional.default.dateString;
const char = import_fictional.default.char;
const digit = import_fictional.default.char.digit;
const word = import_fictional.default.word.options({ unicode: false });
const words = import_fictional.default.words.options({
  minSyllables: 2,
  unicode: false
});
const sentence = import_fictional.default.sentence.options({ unicode: false });
const paragraph = import_fictional.default.paragraph.options({ unicode: false });
const oneOf = import_fictional.default.oneOf;
const times = import_fictional.default.times;
const hex = (input) => int(input, {
  min: 0,
  max: 15
}).toString(16);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  bool,
  char,
  dateString,
  digit,
  float,
  hex,
  int,
  oneOf,
  paragraph,
  sentence,
  times,
  word,
  words
});
