// @hubspot/api-client/lib/codegen/crm/objects/feedback_submissions/types/PromiseAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseSearchApi = exports.PromiseBatchApi = exports.PromiseBasicApi = exports.PromiseAssociationsApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class PromiseAssociationsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableAssociationsApi(configuration, requestFactory, responseProcessor);
    }
    getAll(feedbackSubmissionId, toObjectType, after, limit, _options) {
        const result = this.api.getAll(feedbackSubmissionId, toObjectType, after, limit, _options);
        return result.toPromise();
    }
}
exports.PromiseAssociationsApi = PromiseAssociationsApi;
const ObservableAPI_2 = require("./ObservableAPI");
class PromiseBasicApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_2.ObservableBasicApi(configuration, requestFactory, responseProcessor);
    }
    getById(feedbackSubmissionId, properties, propertiesWithHistory, associations, archived, idProperty, _options) {
        const result = this.api.getById(feedbackSubmissionId, properties, propertiesWithHistory, associations, archived, idProperty, _options);
        return result.toPromise();
    }
    getPage(limit, after, properties, propertiesWithHistory, associations, archived, _options) {
        const result = this.api.getPage(limit, after, properties, propertiesWithHistory, associations, archived, _options);
        return result.toPromise();
    }
}
exports.PromiseBasicApi = PromiseBasicApi;
const ObservableAPI_3 = require("./ObservableAPI");
class PromiseBatchApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_3.ObservableBatchApi(configuration, requestFactory, responseProcessor);
    }
    readBatch(batchReadInputSimplePublicObjectId, archived, _options) {
        const result = this.api.readBatch(batchReadInputSimplePublicObjectId, archived, _options);
        return result.toPromise();
    }
}
exports.PromiseBatchApi = PromiseBatchApi;
const ObservableAPI_4 = require("./ObservableAPI");
class PromiseSearchApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_4.ObservableSearchApi(configuration, requestFactory, responseProcessor);
    }
    doSearch(publicObjectSearchRequest, _options) {
        const result = this.api.doSearch(publicObjectSearchRequest, _options);
        return result.toPromise();
    }
}
exports.PromiseSearchApi = PromiseSearchApi;
