// @hubspot/api-client/lib/codegen/crm/feedback_submissions/models/ObjectSerializer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSerializer = void 0;
__exportStar(require("./AssociatedId"), exports);
__exportStar(require("./BatchInputSimplePublicObjectBatchInput"), exports);
__exportStar(require("./BatchInputSimplePublicObjectId"), exports);
__exportStar(require("./BatchInputSimplePublicObjectInput"), exports);
__exportStar(require("./BatchReadInputSimplePublicObjectId"), exports);
__exportStar(require("./BatchResponseSimplePublicObject"), exports);
__exportStar(require("./BatchResponseSimplePublicObjectWithErrors"), exports);
__exportStar(require("./CollectionResponseAssociatedId"), exports);
__exportStar(require("./CollectionResponseAssociatedIdForwardPaging"), exports);
__exportStar(require("./CollectionResponseSimplePublicObjectWithAssociationsForwardPaging"), exports);
__exportStar(require("./CollectionResponseWithTotalSimplePublicObjectForwardPaging"), exports);
__exportStar(require("./ErrorCategory"), exports);
__exportStar(require("./ErrorDetail"), exports);
__exportStar(require("./Filter"), exports);
__exportStar(require("./FilterGroup"), exports);
__exportStar(require("./ForwardPaging"), exports);
__exportStar(require("./ModelError"), exports);
__exportStar(require("./NextPage"), exports);
__exportStar(require("./Paging"), exports);
__exportStar(require("./PreviousPage"), exports);
__exportStar(require("./PublicObjectSearchRequest"), exports);
__exportStar(require("./SimplePublicObject"), exports);
__exportStar(require("./SimplePublicObjectBatchInput"), exports);
__exportStar(require("./SimplePublicObjectId"), exports);
__exportStar(require("./SimplePublicObjectInput"), exports);
__exportStar(require("./SimplePublicObjectWithAssociations"), exports);
__exportStar(require("./StandardError"), exports);
__exportStar(require("./ValueWithTimestamp"), exports);
const AssociatedId_1 = require("./AssociatedId");
const BatchInputSimplePublicObjectBatchInput_1 = require("./BatchInputSimplePublicObjectBatchInput");
const BatchInputSimplePublicObjectId_1 = require("./BatchInputSimplePublicObjectId");
const BatchInputSimplePublicObjectInput_1 = require("./BatchInputSimplePublicObjectInput");
const BatchReadInputSimplePublicObjectId_1 = require("./BatchReadInputSimplePublicObjectId");
const BatchResponseSimplePublicObject_1 = require("./BatchResponseSimplePublicObject");
const BatchResponseSimplePublicObjectWithErrors_1 = require("./BatchResponseSimplePublicObjectWithErrors");
const CollectionResponseAssociatedId_1 = require("./CollectionResponseAssociatedId");
const CollectionResponseAssociatedIdForwardPaging_1 = require("./CollectionResponseAssociatedIdForwardPaging");
const CollectionResponseSimplePublicObjectWithAssociationsForwardPaging_1 = require("./CollectionResponseSimplePublicObjectWithAssociationsForwardPaging");
const CollectionResponseWithTotalSimplePublicObjectForwardPaging_1 = require("./CollectionResponseWithTotalSimplePublicObjectForwardPaging");
const ErrorCategory_1 = require("./ErrorCategory");
const ErrorDetail_1 = require("./ErrorDetail");
const Filter_1 = require("./Filter");
const FilterGroup_1 = require("./FilterGroup");
const ForwardPaging_1 = require("./ForwardPaging");
const ModelError_1 = require("./ModelError");
const NextPage_1 = require("./NextPage");
const Paging_1 = require("./Paging");
const PreviousPage_1 = require("./PreviousPage");
const PublicObjectSearchRequest_1 = require("./PublicObjectSearchRequest");
const SimplePublicObject_1 = require("./SimplePublicObject");
const SimplePublicObjectBatchInput_1 = require("./SimplePublicObjectBatchInput");
const SimplePublicObjectId_1 = require("./SimplePublicObjectId");
const SimplePublicObjectInput_1 = require("./SimplePublicObjectInput");
const SimplePublicObjectWithAssociations_1 = require("./SimplePublicObjectWithAssociations");
const StandardError_1 = require("./StandardError");
const ValueWithTimestamp_1 = require("./ValueWithTimestamp");
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
const supportedMediaTypes = {
    "application/json": Infinity,
    "application/octet-stream": 0,
    "application/x-www-form-urlencoded": 0
};
let enumsMap = new Set([
    "BatchResponseSimplePublicObjectStatusEnum",
    "BatchResponseSimplePublicObjectWithErrorsStatusEnum",
    "ErrorCategoryHttpStatusEnum",
    "FilterOperatorEnum",
]);
let typeMap = {
    "AssociatedId": AssociatedId_1.AssociatedId,
    "BatchInputSimplePublicObjectBatchInput": BatchInputSimplePublicObjectBatchInput_1.BatchInputSimplePublicObjectBatchInput,
    "BatchInputSimplePublicObjectId": BatchInputSimplePublicObjectId_1.BatchInputSimplePublicObjectId,
    "BatchInputSimplePublicObjectInput": BatchInputSimplePublicObjectInput_1.BatchInputSimplePublicObjectInput,
    "BatchReadInputSimplePublicObjectId": BatchReadInputSimplePublicObjectId_1.BatchReadInputSimplePublicObjectId,
    "BatchResponseSimplePublicObject": BatchResponseSimplePublicObject_1.BatchResponseSimplePublicObject,
    "BatchResponseSimplePublicObjectWithErrors": BatchResponseSimplePublicObjectWithErrors_1.BatchResponseSimplePublicObjectWithErrors,
    "CollectionResponseAssociatedId": CollectionResponseAssociatedId_1.CollectionResponseAssociatedId,
    "CollectionResponseAssociatedIdForwardPaging": CollectionResponseAssociatedIdForwardPaging_1.CollectionResponseAssociatedIdForwardPaging,
    "CollectionResponseSimplePublicObjectWithAssociationsForwardPaging": CollectionResponseSimplePublicObjectWithAssociationsForwardPaging_1.CollectionResponseSimplePublicObjectWithAssociationsForwardPaging,
    "CollectionResponseWithTotalSimplePublicObjectForwardPaging": CollectionResponseWithTotalSimplePublicObjectForwardPaging_1.CollectionResponseWithTotalSimplePublicObjectForwardPaging,
    "ErrorCategory": ErrorCategory_1.ErrorCategory,
    "ErrorDetail": ErrorDetail_1.ErrorDetail,
    "Filter": Filter_1.Filter,
    "FilterGroup": FilterGroup_1.FilterGroup,
    "ForwardPaging": ForwardPaging_1.ForwardPaging,
    "ModelError": ModelError_1.ModelError,
    "NextPage": NextPage_1.NextPage,
    "Paging": Paging_1.Paging,
    "PreviousPage": PreviousPage_1.PreviousPage,
    "PublicObjectSearchRequest": PublicObjectSearchRequest_1.PublicObjectSearchRequest,
    "SimplePublicObject": SimplePublicObject_1.SimplePublicObject,
    "SimplePublicObjectBatchInput": SimplePublicObjectBatchInput_1.SimplePublicObjectBatchInput,
    "SimplePublicObjectId": SimplePublicObjectId_1.SimplePublicObjectId,
    "SimplePublicObjectInput": SimplePublicObjectInput_1.SimplePublicObjectInput,
    "SimplePublicObjectWithAssociations": SimplePublicObjectWithAssociations_1.SimplePublicObjectWithAssociations,
    "StandardError": StandardError_1.StandardError,
    "ValueWithTimestamp": ValueWithTimestamp_1.ValueWithTimestamp,
};
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap.has(expectedType)) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType;
            }
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType;
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    if (typeMap[discriminatorType]) {
                        return discriminatorType;
                    }
                    else {
                        return expectedType;
                    }
                }
                else {
                    return expectedType;
                }
            }
        }
    }
    static serialize(data, type, format) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) {
            let subType = type.replace("Array<", "");
            subType = subType.substring(0, subType.length - 1);
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.serialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type === "Date") {
            if (format == "date") {
                let month = data.getMonth() + 1;
                month = month < 10 ? "0" + month.toString() : month.toString();
                let day = data.getDate();
                day = day < 10 ? "0" + day.toString() : day.toString();
                return data.getFullYear() + "-" + month + "-" + day;
            }
            else {
                return data.toISOString();
            }
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) {
                return data;
            }
            type = this.findCorrectType(data, type);
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static deserialize(data, type, format) {
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) {
            let subType = type.replace("Array<", "");
            subType = subType.substring(0, subType.length - 1);
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.deserialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) {
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.name] = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static normalizeMediaType(mediaType) {
        if (mediaType === undefined) {
            return undefined;
        }
        return mediaType.split(";")[0].trim().toLowerCase();
    }
    static getPreferredMediaType(mediaTypes) {
        if (!mediaTypes) {
            return "application/json";
        }
        const normalMediaTypes = mediaTypes.map(this.normalizeMediaType);
        let selectedMediaType = undefined;
        let selectedRank = -Infinity;
        for (const mediaType of normalMediaTypes) {
            if (supportedMediaTypes[mediaType] > selectedRank) {
                selectedMediaType = mediaType;
                selectedRank = supportedMediaTypes[mediaType];
            }
        }
        if (selectedMediaType === undefined) {
            throw new Error("None of the given media types are supported: " + mediaTypes.join(", "));
        }
        return selectedMediaType;
    }
    static stringify(data, mediaType) {
        if (mediaType === "application/json") {
            return JSON.stringify(data);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.stringify.");
    }
    static parse(rawData, mediaType) {
        if (mediaType === undefined) {
            throw new Error("Cannot parse content. No Content-Type defined.");
        }
        if (mediaType === "application/json") {
            return JSON.parse(rawData);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.parse.");
    }
}
exports.ObjectSerializer = ObjectSerializer;
