// @hubspot/api-client/lib/src/discovery/crm/extensions/accounting/AccountingDiscovery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountingDiscovery = void 0;
const configuration_1 = require("../../../../../codegen/crm/extensions/accounting/configuration");
const index_1 = require("../../../../../codegen/crm/extensions/accounting/index");
const rxjsStub_1 = require("../../../../../codegen/crm/extensions/accounting/rxjsStub");
const ApiClientConfigurator_1 = require("../../../../configuration/ApiClientConfigurator");
class AccountingDiscovery {
    constructor(config) {
        const configuration = configuration_1.createConfiguration(ApiClientConfigurator_1.ApiClientConfigurator.getParams(config, index_1.ServerConfiguration, rxjsStub_1.Observable, rxjsStub_1.Observable));
        this.callbacksApi = new index_1.CallbacksApi(configuration);
        this.invoiceApi = new index_1.InvoiceApi(configuration);
        this.settingsApi = new index_1.SettingsApi(configuration);
        this.syncApi = new index_1.SyncApi(configuration);
        this.userAccountsApi = new index_1.UserAccountsApi(configuration);
    }
}
exports.AccountingDiscovery = AccountingDiscovery;
