// @hubspot/api-client/lib/src/discovery/marketing/events/EventsDiscovery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsDiscovery = void 0;
const configuration_1 = require("../../../../codegen/marketing/events/configuration");
const index_1 = require("../../../../codegen/marketing/events/index");
const rxjsStub_1 = require("../../../../codegen/marketing/events/rxjsStub");
const ApiClientConfigurator_1 = require("../../../configuration/ApiClientConfigurator");
class EventsDiscovery {
    constructor(config) {
        const configuration = configuration_1.createConfiguration(ApiClientConfigurator_1.ApiClientConfigurator.getParams(config, index_1.ServerConfiguration, rxjsStub_1.Observable, rxjsStub_1.Observable));
        this.basicApi = new index_1.BasicApi(configuration);
        this.batchApi = new index_1.BatchApi(configuration);
        this.searchApi = new index_1.SearchApi(configuration);
        this.settingsApi = new index_1.SettingsApi(configuration);
        this.subscriberStateChangesApi = new index_1.SubscriberStateChangesApi(configuration);
    }
}
exports.EventsDiscovery = EventsDiscovery;
