// svix/dist/openapi/types/PromiseAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseStatisticsApi = exports.PromiseOrganizationSettingsApi = exports.PromiseOrganizationApi = exports.PromiseMessageAttemptApi = exports.PromiseMessageApi = exports.PromiseHealthApi = exports.PromiseEventTypeApi = exports.PromiseEndpointApi = exports.PromiseAuthenticationApi = exports.PromiseApplicationApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class PromiseApplicationApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableApplicationApi(configuration, requestFactory, responseProcessor);
    }
    createApplicationApiV1AppPost(applicationIn, _options) {
        const result = this.api.createApplicationApiV1AppPost(applicationIn, _options);
        return result.toPromise();
    }
    deleteApplicationApiV1AppAppIdDelete(appId, _options) {
        const result = this.api.deleteApplicationApiV1AppAppIdDelete(appId, _options);
        return result.toPromise();
    }
    getApplicationApiV1AppAppIdGet(appId, _options) {
        const result = this.api.getApplicationApiV1AppAppIdGet(appId, _options);
        return result.toPromise();
    }
    listApplicationsApiV1AppGet(iterator, limit, _options) {
        const result = this.api.listApplicationsApiV1AppGet(iterator, limit, _options);
        return result.toPromise();
    }
    updateApplicationApiV1AppAppIdPut(appId, applicationIn, _options) {
        const result = this.api.updateApplicationApiV1AppAppIdPut(appId, applicationIn, _options);
        return result.toPromise();
    }
}
exports.PromiseApplicationApi = PromiseApplicationApi;
const ObservableAPI_2 = require("./ObservableAPI");
class PromiseAuthenticationApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_2.ObservableAuthenticationApi(configuration, requestFactory, responseProcessor);
    }
    getDashboardAccessApiV1AuthDashboardAccessAppIdPost(appId, _options) {
        const result = this.api.getDashboardAccessApiV1AuthDashboardAccessAppIdPost(appId, _options);
        return result.toPromise();
    }
    logoutApiV1AuthLogoutPost(_options) {
        const result = this.api.logoutApiV1AuthLogoutPost(_options);
        return result.toPromise();
    }
}
exports.PromiseAuthenticationApi = PromiseAuthenticationApi;
const ObservableAPI_3 = require("./ObservableAPI");
class PromiseEndpointApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_3.ObservableEndpointApi(configuration, requestFactory, responseProcessor);
    }
    createEndpointApiV1AppAppIdEndpointPost(appId, endpointIn, _options) {
        const result = this.api.createEndpointApiV1AppAppIdEndpointPost(appId, endpointIn, _options);
        return result.toPromise();
    }
    deleteEndpointApiV1AppAppIdEndpointEndpointIdDelete(endpointId, appId, _options) {
        const result = this.api.deleteEndpointApiV1AppAppIdEndpointEndpointIdDelete(endpointId, appId, _options);
        return result.toPromise();
    }
    getEndpointApiV1AppAppIdEndpointEndpointIdGet(endpointId, appId, _options) {
        const result = this.api.getEndpointApiV1AppAppIdEndpointEndpointIdGet(endpointId, appId, _options);
        return result.toPromise();
    }
    getEndpointHeadersApiV1AppAppIdEndpointEndpointIdHeadersGet(endpointId, appId, _options) {
        const result = this.api.getEndpointHeadersApiV1AppAppIdEndpointEndpointIdHeadersGet(endpointId, appId, _options);
        return result.toPromise();
    }
    getEndpointSecretApiV1AppAppIdEndpointEndpointIdSecretGet(endpointId, appId, _options) {
        const result = this.api.getEndpointSecretApiV1AppAppIdEndpointEndpointIdSecretGet(endpointId, appId, _options);
        return result.toPromise();
    }
    getEndpointStatsApiV1AppAppIdEndpointEndpointIdStatsGet(endpointId, appId, _options) {
        const result = this.api.getEndpointStatsApiV1AppAppIdEndpointEndpointIdStatsGet(endpointId, appId, _options);
        return result.toPromise();
    }
    listEndpointsApiV1AppAppIdEndpointGet(appId, iterator, limit, _options) {
        const result = this.api.listEndpointsApiV1AppAppIdEndpointGet(appId, iterator, limit, _options);
        return result.toPromise();
    }
    patchEndpointHeadersApiV1AppAppIdEndpointEndpointIdHeadersPatch(appId, endpointId, endpointHeadersIn, _options) {
        const result = this.api.patchEndpointHeadersApiV1AppAppIdEndpointEndpointIdHeadersPatch(appId, endpointId, endpointHeadersIn, _options);
        return result.toPromise();
    }
    recoverFailedWebhooksApiV1AppAppIdEndpointEndpointIdRecoverPost(appId, endpointId, recoverIn, _options) {
        const result = this.api.recoverFailedWebhooksApiV1AppAppIdEndpointEndpointIdRecoverPost(appId, endpointId, recoverIn, _options);
        return result.toPromise();
    }
    rotateEndpointSecretApiV1AppAppIdEndpointEndpointIdSecretRotatePost(endpointId, appId, endpointSecretRotateIn, _options) {
        const result = this.api.rotateEndpointSecretApiV1AppAppIdEndpointEndpointIdSecretRotatePost(endpointId, appId, endpointSecretRotateIn, _options);
        return result.toPromise();
    }
    sendEventTypeExampleMessageApiV1AppAppIdEndpointEndpointIdSendExamplePost(appId, endpointId, eventExampleIn, _options) {
        const result = this.api.sendEventTypeExampleMessageApiV1AppAppIdEndpointEndpointIdSendExamplePost(appId, endpointId, eventExampleIn, _options);
        return result.toPromise();
    }
    updateEndpointApiV1AppAppIdEndpointEndpointIdPut(endpointId, appId, endpointUpdate, _options) {
        const result = this.api.updateEndpointApiV1AppAppIdEndpointEndpointIdPut(endpointId, appId, endpointUpdate, _options);
        return result.toPromise();
    }
    updateEndpointHeadersApiV1AppAppIdEndpointEndpointIdHeadersPut(appId, endpointId, endpointHeadersIn, _options) {
        const result = this.api.updateEndpointHeadersApiV1AppAppIdEndpointEndpointIdHeadersPut(appId, endpointId, endpointHeadersIn, _options);
        return result.toPromise();
    }
}
exports.PromiseEndpointApi = PromiseEndpointApi;
const ObservableAPI_4 = require("./ObservableAPI");
class PromiseEventTypeApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_4.ObservableEventTypeApi(configuration, requestFactory, responseProcessor);
    }
    createEventTypeApiV1EventTypePost(eventTypeIn, _options) {
        const result = this.api.createEventTypeApiV1EventTypePost(eventTypeIn, _options);
        return result.toPromise();
    }
    deleteEventTypeApiV1EventTypeEventTypeNameDelete(eventTypeName, _options) {
        const result = this.api.deleteEventTypeApiV1EventTypeEventTypeNameDelete(eventTypeName, _options);
        return result.toPromise();
    }
    generateSchemaExampleApiV1EventTypeSchemaGenerateExamplePost(eventTypeSchemaIn, _options) {
        const result = this.api.generateSchemaExampleApiV1EventTypeSchemaGenerateExamplePost(eventTypeSchemaIn, _options);
        return result.toPromise();
    }
    getEventTypeApiV1EventTypeEventTypeNameGet(eventTypeName, _options) {
        const result = this.api.getEventTypeApiV1EventTypeEventTypeNameGet(eventTypeName, _options);
        return result.toPromise();
    }
    listEventTypesApiV1EventTypeGet(iterator, limit, withContent, includeArchived, _options) {
        const result = this.api.listEventTypesApiV1EventTypeGet(iterator, limit, withContent, includeArchived, _options);
        return result.toPromise();
    }
    updateEventTypeApiV1EventTypeEventTypeNamePut(eventTypeName, eventTypeUpdate, _options) {
        const result = this.api.updateEventTypeApiV1EventTypeEventTypeNamePut(eventTypeName, eventTypeUpdate, _options);
        return result.toPromise();
    }
}
exports.PromiseEventTypeApi = PromiseEventTypeApi;
const ObservableAPI_5 = require("./ObservableAPI");
class PromiseHealthApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_5.ObservableHealthApi(configuration, requestFactory, responseProcessor);
    }
    healthApiV1HealthGet(_options) {
        const result = this.api.healthApiV1HealthGet(_options);
        return result.toPromise();
    }
}
exports.PromiseHealthApi = PromiseHealthApi;
const ObservableAPI_6 = require("./ObservableAPI");
class PromiseMessageApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_6.ObservableMessageApi(configuration, requestFactory, responseProcessor);
    }
    createMessageApiV1AppAppIdMsgPost(appId, messageIn, withContent, _options) {
        const result = this.api.createMessageApiV1AppAppIdMsgPost(appId, messageIn, withContent, _options);
        return result.toPromise();
    }
    getMessageApiV1AppAppIdMsgMsgIdGet(msgId, appId, _options) {
        const result = this.api.getMessageApiV1AppAppIdMsgMsgIdGet(msgId, appId, _options);
        return result.toPromise();
    }
    listMessagesApiV1AppAppIdMsgGet(appId, iterator, limit, eventTypes, before, _options) {
        const result = this.api.listMessagesApiV1AppAppIdMsgGet(appId, iterator, limit, eventTypes, before, _options);
        return result.toPromise();
    }
}
exports.PromiseMessageApi = PromiseMessageApi;
const ObservableAPI_7 = require("./ObservableAPI");
class PromiseMessageAttemptApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_7.ObservableMessageAttemptApi(configuration, requestFactory, responseProcessor);
    }
    getAttemptApiV1AppAppIdMsgMsgIdAttemptAttemptIdGet(attemptId, msgId, appId, _options) {
        const result = this.api.getAttemptApiV1AppAppIdMsgMsgIdAttemptAttemptIdGet(attemptId, msgId, appId, _options);
        return result.toPromise();
    }
    listAttemptedDestinationsApiV1AppAppIdMsgMsgIdEndpointGet(msgId, appId, iterator, limit, _options) {
        const result = this.api.listAttemptedDestinationsApiV1AppAppIdMsgMsgIdEndpointGet(msgId, appId, iterator, limit, _options);
        return result.toPromise();
    }
    listAttemptedMessagesApiV1AppAppIdEndpointEndpointIdMsgGet(endpointId, appId, iterator, limit, status, before, _options) {
        const result = this.api.listAttemptedMessagesApiV1AppAppIdEndpointEndpointIdMsgGet(endpointId, appId, iterator, limit, status, before, _options);
        return result.toPromise();
    }
    listAttemptsApiV1AppAppIdMsgMsgIdAttemptGet(appId, msgId, iterator, limit, status, _options) {
        const result = this.api.listAttemptsApiV1AppAppIdMsgMsgIdAttemptGet(appId, msgId, iterator, limit, status, _options);
        return result.toPromise();
    }
    listAttemptsForEndpointApiV1AppAppIdMsgMsgIdEndpointEndpointIdAttemptGet(msgId, appId, endpointId, iterator, limit, eventTypes, status, before, _options) {
        const result = this.api.listAttemptsForEndpointApiV1AppAppIdMsgMsgIdEndpointEndpointIdAttemptGet(msgId, appId, endpointId, iterator, limit, eventTypes, status, before, _options);
        return result.toPromise();
    }
    resendWebhookApiV1AppAppIdMsgMsgIdEndpointEndpointIdResendPost(endpointId, msgId, appId, _options) {
        const result = this.api.resendWebhookApiV1AppAppIdMsgMsgIdEndpointEndpointIdResendPost(endpointId, msgId, appId, _options);
        return result.toPromise();
    }
}
exports.PromiseMessageAttemptApi = PromiseMessageAttemptApi;
const ObservableAPI_8 = require("./ObservableAPI");
class PromiseOrganizationApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_8.ObservableOrganizationApi(configuration, requestFactory, responseProcessor);
    }
    exportOrganizationConfigurationApiV1OrgExportPost(body, _options) {
        const result = this.api.exportOrganizationConfigurationApiV1OrgExportPost(body, _options);
        return result.toPromise();
    }
    importOrganizationConfigurationApiV1OrgImportPost(importOrganizationIn, _options) {
        const result = this.api.importOrganizationConfigurationApiV1OrgImportPost(importOrganizationIn, _options);
        return result.toPromise();
    }
}
exports.PromiseOrganizationApi = PromiseOrganizationApi;
const ObservableAPI_9 = require("./ObservableAPI");
class PromiseOrganizationSettingsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_9.ObservableOrganizationSettingsApi(configuration, requestFactory, responseProcessor);
    }
    getOrgSettingsApiV1OrgSettingsGet(_options) {
        const result = this.api.getOrgSettingsApiV1OrgSettingsGet(_options);
        return result.toPromise();
    }
}
exports.PromiseOrganizationSettingsApi = PromiseOrganizationSettingsApi;
const ObservableAPI_10 = require("./ObservableAPI");
class PromiseStatisticsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_10.ObservableStatisticsApi(configuration, requestFactory, responseProcessor);
    }
    getAppAttemptStatsApiV1StatsAppAppIdAttemptGet(appId, startDate, endDate, _options) {
        const result = this.api.getAppAttemptStatsApiV1StatsAppAppIdAttemptGet(appId, startDate, endDate, _options);
        return result.toPromise();
    }
    getEpStatsApiV1StatsAppAppIdEpEndpointIdAttemptGet(endpointId, appId, startDate, endDate, _options) {
        const result = this.api.getEpStatsApiV1StatsAppAppIdEpEndpointIdAttemptGet(endpointId, appId, startDate, endDate, _options);
        return result.toPromise();
    }
}
exports.PromiseStatisticsApi = PromiseStatisticsApi;
