// @tryvital/vital-node/dist/client/models/activity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RespiratoryAllergen;
(function (RespiratoryAllergen) {
    RespiratoryAllergen["D_PTERONYSSINUS"] = "d_pteronyssinus";
    RespiratoryAllergen["D_FARINAE"] = "d_farinae";
    RespiratoryAllergen["CAT_DANDER"] = "cat_dander";
    RespiratoryAllergen["DOG_DANDER"] = "dog_dander";
    RespiratoryAllergen["HORSE_DANDER"] = "horse_dander";
    RespiratoryAllergen["MOUSE_URINE"] = "mouse_urine";
    RespiratoryAllergen["COCKROACH"] = "cockroach";
    RespiratoryAllergen["ALTERNARIA_ALTERNATA"] = "alternaria_alternata";
    RespiratoryAllergen["ASPERGILLUS_FUMIGATUS"] = "aspergillus_fumigatus";
    RespiratoryAllergen["CLADOSPORIUM_HERBARUM"] = "cladosporium_herbarum";
    RespiratoryAllergen["PENICILLIUM_NOTATUM"] = "penicillium_notatum";
    RespiratoryAllergen["AUREOBASIDIUM_PULLULANS"] = "aureobasidium_pullulans";
    RespiratoryAllergen["BAHIA_GRASS"] = "bahia_grass";
    RespiratoryAllergen["BERMUDA_GRASS"] = "bermuda_grass";
    RespiratoryAllergen["JOHNSON_GRASS"] = "johnson_grass";
    RespiratoryAllergen["PERENNIAL_RYEGRASS"] = "perennial_ryegrass";
    RespiratoryAllergen["TIMOTHY_GRASS"] = "timothy_grass";
    RespiratoryAllergen["ACACIA"] = "acacia";
    RespiratoryAllergen["ALDER"] = "alder";
    RespiratoryAllergen["AUSTRALIAN_PINE"] = "australian_pine";
    RespiratoryAllergen["BIRCH"] = "birch";
    RespiratoryAllergen["COTTONWOOD"] = "cottonwood";
    RespiratoryAllergen["ELM"] = "elm";
    RespiratoryAllergen["MAPLE_BOX_ELDER"] = "maple_box_elder";
    RespiratoryAllergen["MAPLE_LEAF_SYCAMORE"] = "maple_leaf_sycamore";
    RespiratoryAllergen["MOUNTAIN_CEDAR"] = "mountain_cedar";
    RespiratoryAllergen["MULBERRY"] = "mulberry";
    RespiratoryAllergen["OAK"] = "oak";
    RespiratoryAllergen["OLIVE"] = "olive";
    RespiratoryAllergen["PECAN_HICKORY"] = "pecan_hickory";
    RespiratoryAllergen["WALNUT"] = "walnut";
    RespiratoryAllergen["WHITE_ASH"] = "white_ash";
    RespiratoryAllergen["COMMON_RAGWEED"] = "common_ragweed";
    RespiratoryAllergen["MUGWORT"] = "mugwort";
    RespiratoryAllergen["NETTLE"] = "nettle";
    RespiratoryAllergen["ROUGH_MARSH_ELDER"] = "rough_marsh_elder";
    RespiratoryAllergen["ROUGH_PIGWEED"] = "rough_pigweed";
    RespiratoryAllergen["RUSSIAN_THISTLE"] = "russian_thistle";
    RespiratoryAllergen["SHEEP_SORREL"] = "sheep_sorrel";
    RespiratoryAllergen["BLOMIA_TROPICALIS"] = "blomia_tropicalis";
})(RespiratoryAllergen || (RespiratoryAllergen = {}));
var FoodAllergen;
(function (FoodAllergen) {
    FoodAllergen["CHEDDAR_CHEESE"] = "cheddar_cheese";
    FoodAllergen["COTTAGE_CHEESE"] = "cottage_cheese";
    FoodAllergen["COW_MILK"] = "cow_milk";
    FoodAllergen["MOZZARELLA_CHEESE"] = "mozzarella_cheese";
    FoodAllergen["YOGURT"] = "yogurt";
    FoodAllergen["EGG_WHITE"] = "egg_white";
    FoodAllergen["EGG_YOLK"] = "egg_yolk";
    FoodAllergen["APPLE"] = "apple";
    FoodAllergen["AVOCADO"] = "avocado";
    FoodAllergen["BANANA"] = "banana";
    FoodAllergen["BLUEBERRY"] = "blueberry";
    FoodAllergen["CANTALOUPE"] = "cantaloupe";
    FoodAllergen["COCONUT"] = "coconut";
    FoodAllergen["GRAPE"] = "grape";
    FoodAllergen["GRAPEFRUIT"] = "grapefruit";
    FoodAllergen["LEMON"] = "lemon";
    FoodAllergen["ORANGE"] = "orange";
    FoodAllergen["PEACH"] = "peach";
    FoodAllergen["PEAR"] = "pear";
    FoodAllergen["PINEAPPLE"] = "pineapple";
    FoodAllergen["STRAWBERRY"] = "strawberry";
    FoodAllergen["TOMATO"] = "tomato";
    FoodAllergen["WATERMELON"] = "watermelon";
    FoodAllergen["BAKERS_YEAST"] = "bakers_yeast";
    FoodAllergen["BARLEY"] = "barley";
    FoodAllergen["BRAN"] = "bran";
    FoodAllergen["BREWERS_YEAST"] = "brewers_yeast";
    FoodAllergen["BROWN_RICE"] = "brown_rice";
    FoodAllergen["GLUTEN"] = "gluten";
    FoodAllergen["MALT"] = "malt";
    FoodAllergen["OATS"] = "oats";
    FoodAllergen["RYE"] = "rye";
    FoodAllergen["WHEAT"] = "wheat";
    FoodAllergen["GREEN_BEAN"] = "green_bean";
    FoodAllergen["GREEN_PEA"] = "green_pea";
    FoodAllergen["LIMA_BEAN"] = "lima_bean";
    FoodAllergen["PEANUT"] = "peanut";
    FoodAllergen["SOYBEAN"] = "soybean";
    FoodAllergen["BEEF"] = "beef";
    FoodAllergen["CHICKEN"] = "chicken";
    FoodAllergen["LAMB"] = "lamb";
    FoodAllergen["PORK"] = "pork";
    FoodAllergen["TURKEY"] = "turkey";
    FoodAllergen["CLAM"] = "clam";
    FoodAllergen["CODFISH"] = "codfish";
    FoodAllergen["CRAB"] = "crab";
    FoodAllergen["HADDOCK"] = "haddock";
    FoodAllergen["LOBSTER"] = "lobster";
    FoodAllergen["PRAWN"] = "prawn";
    FoodAllergen["SALMON"] = "salmon";
    FoodAllergen["SCALLOP"] = "scallop";
    FoodAllergen["SOLE"] = "sole";
    FoodAllergen["SWORDFISH"] = "swordfish";
    FoodAllergen["TUNA"] = "tuna";
    FoodAllergen["ALMOND"] = "almond";
    FoodAllergen["BLACK_WALNUT"] = "black_walnut";
    FoodAllergen["CASHEW"] = "cashew";
    FoodAllergen["CHIA_SEED"] = "chia_seed";
    FoodAllergen["SAFFLOWER"] = "safflower";
    FoodAllergen["SESAME"] = "sesame";
    FoodAllergen["SUNFLOWER"] = "sunflower";
    FoodAllergen["ASPARAGUS"] = "asparagus";
    FoodAllergen["BELL_PEPPER"] = "bell_pepper";
    FoodAllergen["BROCCOLI"] = "broccoli";
    FoodAllergen["CABBAGE"] = "cabbage";
    FoodAllergen["CARROT"] = "carrot";
    FoodAllergen["CAULIFLOWER"] = "cauliflower";
    FoodAllergen["CELERY"] = "celery";
    FoodAllergen["CORN"] = "corn";
    FoodAllergen["CUCUMBER"] = "cucumber";
    FoodAllergen["EGGPLANT"] = "eggplant";
    FoodAllergen["GREEN_OLIVE"] = "green_olive";
    FoodAllergen["KALE"] = "kale";
    FoodAllergen["KELP"] = "kelp";
    FoodAllergen["LETTUCE"] = "lettuce";
    FoodAllergen["MUSHROOM"] = "mushroom";
    FoodAllergen["ONION"] = "onion";
    FoodAllergen["POTATO"] = "potato";
    FoodAllergen["SPINACH"] = "spinach";
    FoodAllergen["SQUASH"] = "squash";
    FoodAllergen["SWEET_POTATO"] = "sweet_potato";
    FoodAllergen["BASIL"] = "basil";
    FoodAllergen["BAY_LEAF"] = "bay_leaf";
    FoodAllergen["BLACK_PEPPER"] = "black_pepper";
    FoodAllergen["BLACK_TEA"] = "black_tea";
    FoodAllergen["COCOA"] = "cocoa";
    FoodAllergen["COFFEE"] = "coffee";
    FoodAllergen["COLA"] = "cola";
    FoodAllergen["CINNAMON"] = "cinnamon";
    FoodAllergen["DILL"] = "dill";
    FoodAllergen["GARLIC"] = "garlic";
    FoodAllergen["GINGER"] = "ginger";
    FoodAllergen["HONEY"] = "honey";
    FoodAllergen["MUSTARD"] = "mustard";
    FoodAllergen["OREGANO"] = "oregano";
    FoodAllergen["TARRAGON"] = "tarragon";
})(FoodAllergen || (FoodAllergen = {}));
