// @hubspot/api-client/lib/src/discovery/crm/objects/ObjectsDiscovery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectsDiscovery = void 0;
const configuration_1 = require("../../../../codegen/crm/objects/configuration");
const index_1 = require("../../../../codegen/crm/objects/index");
const rxjsStub_1 = require("../../../../codegen/crm/objects/rxjsStub");
const ApiClientConfigurator_1 = require("../../../configuration/ApiClientConfigurator");
const CallsDiscovery_1 = require("./calls/CallsDiscovery");
const EmailsDiscovery_1 = require("./emails/EmailsDiscovery");
const FeedbackSubmissionsDiscovery_1 = require("./feedback_submissions/FeedbackSubmissionsDiscovery");
const MeetingsDiscovery_1 = require("./meetings/MeetingsDiscovery");
const NotesDiscovery_1 = require("./notes/NotesDiscovery");
const TasksDiscovery_1 = require("./tasks/TasksDiscovery");
class ObjectsDiscovery {
    constructor(config) {
        const configuration = configuration_1.createConfiguration(ApiClientConfigurator_1.ApiClientConfigurator.getParams(config, index_1.ServerConfiguration, rxjsStub_1.Observable, rxjsStub_1.Observable));
        this.associationsApi = new index_1.AssociationsApi(configuration);
        this.basicApi = new index_1.BasicApi(configuration);
        this.batchApi = new index_1.BatchApi(configuration);
        this.calls = new CallsDiscovery_1.CallsDiscovery(config);
        this.emails = new EmailsDiscovery_1.EmailsDiscovery(config);
        this.gdprApi = new index_1.GDPRApi(configuration);
        this.feedbackSubmissions = new FeedbackSubmissionsDiscovery_1.FeedbackSubmissionsDiscovery(config);
        this.meetings = new MeetingsDiscovery_1.MeetingsDiscovery(config);
        this.notes = new NotesDiscovery_1.NotesDiscovery(config);
        this.publicObjectApi = new index_1.PublicObjectApi(configuration);
        this.tasks = new TasksDiscovery_1.TasksDiscovery(config);
        this.searchApi = new index_1.SearchApi(configuration);
    }
}
exports.ObjectsDiscovery = ObjectsDiscovery;
