// slack-block-builder/dist/surfaces/workflow-step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowStepBuilder = void 0;
const base_1 = require("../internal/base");
const constants_1 = require("../internal/constants");
const dto_1 = require("../internal/dto");
const helpers_1 = require("../internal/helpers");
const methods_1 = require("../internal/methods");
/**
 * @@link https://api.slack.com/reference/workflows/configuration-view
 * @@displayName Workflow Step
 */
class WorkflowStepBuilder extends base_1.SurfaceBuilderBase {
    /** @internal */
    build() {
        return this.getResult(dto_1.SlackWorkflowStepDto, {
            type: constants_1.SurfaceType.WorkflowStep,
            title: helpers_1.getPlainTextObject(this.props.title),
            blocks: helpers_1.getBuilderResults(this.props.blocks),
            close: helpers_1.getPlainTextObject(this.props.close),
            submit: helpers_1.getPlainTextObject(this.props.submit),
        });
    }
}
exports.WorkflowStepBuilder = WorkflowStepBuilder;
helpers_1.applyMixins(WorkflowStepBuilder, [
    methods_1.Blocks,
    methods_1.CallbackId,
    methods_1.PrivateMetaData,
    methods_1.SubmitDisabled,
    methods_1.BuildToJSON,
    methods_1.BuildToObject,
    methods_1.GetBlocks,
    methods_1.GetPreviewUrl,
    methods_1.PrintPreviewUrl,
]);
