// @hubspot/api-client/lib/codegen/oauth/types/ObjectParamAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectTokensApi = exports.ObjectRefreshTokensApi = exports.ObjectAccessTokensApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class ObjectAccessTokensApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableAccessTokensApi(configuration, requestFactory, responseProcessor);
    }
    getAccessToken(param, options) {
        return this.api.getAccessToken(param.token, options).toPromise();
    }
}
exports.ObjectAccessTokensApi = ObjectAccessTokensApi;
const ObservableAPI_2 = require("./ObservableAPI");
class ObjectRefreshTokensApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_2.ObservableRefreshTokensApi(configuration, requestFactory, responseProcessor);
    }
    archiveRefreshToken(param, options) {
        return this.api.archiveRefreshToken(param.token, options).toPromise();
    }
    getRefreshToken(param, options) {
        return this.api.getRefreshToken(param.token, options).toPromise();
    }
}
exports.ObjectRefreshTokensApi = ObjectRefreshTokensApi;
const ObservableAPI_3 = require("./ObservableAPI");
class ObjectTokensApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_3.ObservableTokensApi(configuration, requestFactory, responseProcessor);
    }
    createToken(param, options) {
        return this.api.createToken(param.grantType, param.code, param.redirectUri, param.clientId, param.clientSecret, param.refreshToken, options).toPromise();
    }
}
exports.ObjectTokensApi = ObjectTokensApi;
