// @hubspot/api-client/lib/src/discovery/cms/source_code/SourceCodeDiscovery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceCodeDiscovery = void 0;
const configuration_1 = require("../../../../codegen/cms/source_code/configuration");
const index_1 = require("../../../../codegen/cms/source_code/index");
const rxjsStub_1 = require("../../../../codegen/cms/source_code/rxjsStub");
const ApiClientConfigurator_1 = require("../../../configuration/ApiClientConfigurator");
class SourceCodeDiscovery {
    constructor(config) {
        const configuration = configuration_1.createConfiguration(ApiClientConfigurator_1.ApiClientConfigurator.getParams(config, index_1.ServerConfiguration, rxjsStub_1.Observable, rxjsStub_1.Observable));
        this.contentApi = new index_1.ContentApi(configuration);
        this.extractApi = new index_1.ExtractApi(configuration);
        this.metadataApi = new index_1.MetadataApi(configuration);
        this.sourceCodeExtractApi = new index_1.SourceCodeExtractApi(configuration);
        this.validationApi = new index_1.ValidationApi(configuration);
    }
}
exports.SourceCodeDiscovery = SourceCodeDiscovery;
