// slack-block-builder/dist/components/accordion-ui-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccordionUIComponent = void 0;
const blocks_1 = require("../blocks");
const elements_1 = require("../elements");
const constants_1 = require("../internal/constants");
class AccordionUIComponent {
    constructor(params) {
        this.items = params.items;
        this.paginator = params.paginator;
        this.expandButtonText = params.expandButtonText || constants_1.ComponentUIText.More;
        this.collapseButtonText = params.collapseButtonText || constants_1.ComponentUIText.Close;
        this.titleTextFunction = params.titleTextFunction;
        this.actionIdFunction = params.actionIdFunction;
        this.builderFunction = params.builderFunction;
    }
    getBlocks() {
        return this.items.map((item, index) => {
            const isExpanded = this.paginator.checkItemIsExpandedByIndex(index);
            const blocks = [
                blocks_1.Blocks.Section({ text: this.titleTextFunction({ item }) })
                    .accessory(elements_1.Elements.Button({
                    text: isExpanded ? this.collapseButtonText : this.expandButtonText,
                    actionId: this.actionIdFunction({
                        expandedItems: this.paginator.getNextStateByItemIndex(index),
                    }),
                })),
                ...isExpanded ? this.builderFunction({ item }) : [],
            ];
            return index === 0 ? blocks : [blocks_1.Blocks.Divider(), ...blocks];
        }).flat();
    }
}
exports.AccordionUIComponent = AccordionUIComponent;
