// @hubspot/api-client/lib/codegen/crm/products/types/PromiseAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseSearchApi = exports.PromisePublicObjectApi = exports.PromiseBatchApi = exports.PromiseBasicApi = exports.PromiseAssociationsApi = void 0;
const ObservableAPI_1 = require("./ObservableAPI");
class PromiseAssociationsApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_1.ObservableAssociationsApi(configuration, requestFactory, responseProcessor);
    }
    archive(productId, toObjectType, toObjectId, associationType, _options) {
        const result = this.api.archive(productId, toObjectType, toObjectId, associationType, _options);
        return result.toPromise();
    }
    create(productId, toObjectType, toObjectId, associationType, _options) {
        const result = this.api.create(productId, toObjectType, toObjectId, associationType, _options);
        return result.toPromise();
    }
    getAll(productId, toObjectType, after, limit, _options) {
        const result = this.api.getAll(productId, toObjectType, after, limit, _options);
        return result.toPromise();
    }
}
exports.PromiseAssociationsApi = PromiseAssociationsApi;
const ObservableAPI_2 = require("./ObservableAPI");
class PromiseBasicApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_2.ObservableBasicApi(configuration, requestFactory, responseProcessor);
    }
    archive(productId, _options) {
        const result = this.api.archive(productId, _options);
        return result.toPromise();
    }
    create(simplePublicObjectInput, _options) {
        const result = this.api.create(simplePublicObjectInput, _options);
        return result.toPromise();
    }
    getById(productId, properties, propertiesWithHistory, associations, archived, idProperty, _options) {
        const result = this.api.getById(productId, properties, propertiesWithHistory, associations, archived, idProperty, _options);
        return result.toPromise();
    }
    getPage(limit, after, properties, propertiesWithHistory, associations, archived, _options) {
        const result = this.api.getPage(limit, after, properties, propertiesWithHistory, associations, archived, _options);
        return result.toPromise();
    }
    update(productId, simplePublicObjectInput, idProperty, _options) {
        const result = this.api.update(productId, simplePublicObjectInput, idProperty, _options);
        return result.toPromise();
    }
}
exports.PromiseBasicApi = PromiseBasicApi;
const ObservableAPI_3 = require("./ObservableAPI");
class PromiseBatchApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_3.ObservableBatchApi(configuration, requestFactory, responseProcessor);
    }
    archive(batchInputSimplePublicObjectId, _options) {
        const result = this.api.archive(batchInputSimplePublicObjectId, _options);
        return result.toPromise();
    }
    create(batchInputSimplePublicObjectInput, _options) {
        const result = this.api.create(batchInputSimplePublicObjectInput, _options);
        return result.toPromise();
    }
    read(batchReadInputSimplePublicObjectId, archived, _options) {
        const result = this.api.read(batchReadInputSimplePublicObjectId, archived, _options);
        return result.toPromise();
    }
    update(batchInputSimplePublicObjectBatchInput, _options) {
        const result = this.api.update(batchInputSimplePublicObjectBatchInput, _options);
        return result.toPromise();
    }
}
exports.PromiseBatchApi = PromiseBatchApi;
const ObservableAPI_4 = require("./ObservableAPI");
class PromisePublicObjectApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_4.ObservablePublicObjectApi(configuration, requestFactory, responseProcessor);
    }
    merge(publicMergeInput, _options) {
        const result = this.api.merge(publicMergeInput, _options);
        return result.toPromise();
    }
}
exports.PromisePublicObjectApi = PromisePublicObjectApi;
const ObservableAPI_5 = require("./ObservableAPI");
class PromiseSearchApi {
    constructor(configuration, requestFactory, responseProcessor) {
        this.api = new ObservableAPI_5.ObservableSearchApi(configuration, requestFactory, responseProcessor);
    }
    doSearch(publicObjectSearchRequest, _options) {
        const result = this.api.doSearch(publicObjectSearchRequest, _options);
        return result.toPromise();
    }
}
exports.PromiseSearchApi = PromiseSearchApi;
