// @hubspot/api-client/lib/codegen/crm/imports/models/ObjectSerializer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSerializer = void 0;
__exportStar(require("./ActionResponse"), exports);
__exportStar(require("./CollectionResponsePublicImportErrorForwardPaging"), exports);
__exportStar(require("./CollectionResponsePublicImportResponse"), exports);
__exportStar(require("./ErrorDetail"), exports);
__exportStar(require("./ForwardPaging"), exports);
__exportStar(require("./ImportRowCore"), exports);
__exportStar(require("./ModelError"), exports);
__exportStar(require("./NextPage"), exports);
__exportStar(require("./Paging"), exports);
__exportStar(require("./PreviousPage"), exports);
__exportStar(require("./PublicImportError"), exports);
__exportStar(require("./PublicImportMetadata"), exports);
__exportStar(require("./PublicImportResponse"), exports);
__exportStar(require("./PublicObjectListRecord"), exports);
const ActionResponse_1 = require("./ActionResponse");
const CollectionResponsePublicImportErrorForwardPaging_1 = require("./CollectionResponsePublicImportErrorForwardPaging");
const CollectionResponsePublicImportResponse_1 = require("./CollectionResponsePublicImportResponse");
const ErrorDetail_1 = require("./ErrorDetail");
const ForwardPaging_1 = require("./ForwardPaging");
const ImportRowCore_1 = require("./ImportRowCore");
const ModelError_1 = require("./ModelError");
const NextPage_1 = require("./NextPage");
const Paging_1 = require("./Paging");
const PreviousPage_1 = require("./PreviousPage");
const PublicImportError_1 = require("./PublicImportError");
const PublicImportMetadata_1 = require("./PublicImportMetadata");
const PublicImportResponse_1 = require("./PublicImportResponse");
const PublicObjectListRecord_1 = require("./PublicObjectListRecord");
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
const supportedMediaTypes = {
    "application/json": Infinity,
    "application/octet-stream": 0,
    "application/x-www-form-urlencoded": 0
};
let enumsMap = new Set([
    "ActionResponseStatusEnum",
    "PublicImportErrorErrorTypeEnum",
    "PublicImportErrorObjectTypeEnum",
    "PublicImportResponseStateEnum",
]);
let typeMap = {
    "ActionResponse": ActionResponse_1.ActionResponse,
    "CollectionResponsePublicImportErrorForwardPaging": CollectionResponsePublicImportErrorForwardPaging_1.CollectionResponsePublicImportErrorForwardPaging,
    "CollectionResponsePublicImportResponse": CollectionResponsePublicImportResponse_1.CollectionResponsePublicImportResponse,
    "ErrorDetail": ErrorDetail_1.ErrorDetail,
    "ForwardPaging": ForwardPaging_1.ForwardPaging,
    "ImportRowCore": ImportRowCore_1.ImportRowCore,
    "ModelError": ModelError_1.ModelError,
    "NextPage": NextPage_1.NextPage,
    "Paging": Paging_1.Paging,
    "PreviousPage": PreviousPage_1.PreviousPage,
    "PublicImportError": PublicImportError_1.PublicImportError,
    "PublicImportMetadata": PublicImportMetadata_1.PublicImportMetadata,
    "PublicImportResponse": PublicImportResponse_1.PublicImportResponse,
    "PublicObjectListRecord": PublicObjectListRecord_1.PublicObjectListRecord,
};
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap.has(expectedType)) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType;
            }
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType;
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    if (typeMap[discriminatorType]) {
                        return discriminatorType;
                    }
                    else {
                        return expectedType;
                    }
                }
                else {
                    return expectedType;
                }
            }
        }
    }
    static serialize(data, type, format) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) {
            let subType = type.replace("Array<", "");
            subType = subType.substring(0, subType.length - 1);
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.serialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type === "Date") {
            if (format == "date") {
                let month = data.getMonth() + 1;
                month = month < 10 ? "0" + month.toString() : month.toString();
                let day = data.getDate();
                day = day < 10 ? "0" + day.toString() : day.toString();
                return data.getFullYear() + "-" + month + "-" + day;
            }
            else {
                return data.toISOString();
            }
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) {
                return data;
            }
            type = this.findCorrectType(data, type);
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static deserialize(data, type, format) {
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) {
            let subType = type.replace("Array<", "");
            subType = subType.substring(0, subType.length - 1);
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.deserialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) {
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.name] = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static normalizeMediaType(mediaType) {
        if (mediaType === undefined) {
            return undefined;
        }
        return mediaType.split(";")[0].trim().toLowerCase();
    }
    static getPreferredMediaType(mediaTypes) {
        if (!mediaTypes) {
            return "application/json";
        }
        const normalMediaTypes = mediaTypes.map(this.normalizeMediaType);
        let selectedMediaType = undefined;
        let selectedRank = -Infinity;
        for (const mediaType of normalMediaTypes) {
            if (supportedMediaTypes[mediaType] > selectedRank) {
                selectedMediaType = mediaType;
                selectedRank = supportedMediaTypes[mediaType];
            }
        }
        if (selectedMediaType === undefined) {
            throw new Error("None of the given media types are supported: " + mediaTypes.join(", "));
        }
        return selectedMediaType;
    }
    static stringify(data, mediaType) {
        if (mediaType === "application/json") {
            return JSON.stringify(data);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.stringify.");
    }
    static parse(rawData, mediaType) {
        if (mediaType === undefined) {
            throw new Error("Cannot parse content. No Content-Type defined.");
        }
        if (mediaType === "application/json") {
            return JSON.parse(rawData);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.parse.");
    }
}
exports.ObjectSerializer = ObjectSerializer;
