// @boxyhq/saml-jackson/dist/db/redis.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const redis = __importStar(require("redis"));
const dbutils = __importStar(require("./utils"));
class Redis {
    constructor(options) {
        this.options = options;
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = {};
            if (this.options && this.options.url) {
                opts['socket'] = {
                    url: this.options.url,
                };
            }
            this.client = redis.createClient(opts);
            this.client.on('error', (err) => console.log('Redis Client Error', err));
            yield this.client.connect();
            return this;
        });
    }
    get(namespace, key) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield this.client.get(dbutils.key(namespace, key));
            if (res) {
                return JSON.parse(res);
            }
            return null;
        });
    }
    getByIndex(namespace, idx) {
        return __awaiter(this, void 0, void 0, function* () {
            const dbKeys = yield this.client.sMembers(dbutils.keyForIndex(namespace, idx));
            const ret = [];
            for (const dbKey of dbKeys || []) {
                ret.push(yield this.get(namespace, dbKey));
            }
            return ret;
        });
    }
    put(namespace, key, val, ttl = 0, ...indexes) {
        return __awaiter(this, void 0, void 0, function* () {
            let tx = this.client.multi();
            const k = dbutils.key(namespace, key);
            tx = tx.set(k, JSON.stringify(val));
            if (ttl) {
                tx = tx.expire(k, ttl);
            }
            // no ttl support for secondary indexes
            for (const idx of indexes || []) {
                const idxKey = dbutils.keyForIndex(namespace, idx);
                tx = tx.sAdd(idxKey, key);
                tx = tx.sAdd(dbutils.keyFromParts(dbutils.indexPrefix, k), idxKey);
            }
            yield tx.exec();
        });
    }
    delete(namespace, key) {
        return __awaiter(this, void 0, void 0, function* () {
            let tx = this.client.multi();
            const k = dbutils.key(namespace, key);
            tx = tx.del(k);
            const idxKey = dbutils.keyFromParts(dbutils.indexPrefix, k);
            // delete secondary indexes and then the mapping of the seconary indexes
            const dbKeys = yield this.client.sMembers(idxKey);
            for (const dbKey of dbKeys || []) {
                tx.sRem(dbKey, key);
            }
            tx.del(idxKey);
            return yield tx.exec();
        });
    }
}
exports.default = {
    new: (options) => __awaiter(void 0, void 0, void 0, function* () {
        return yield new Redis(options).init();
    }),
};
