// @hubspot/api-client/lib/src/discovery/oauth/OauthDiscovery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OauthDiscovery = void 0;
const _ = require("lodash");
const qs = require("querystring");
const configuration_1 = require("../../../codegen/oauth/configuration");
const index_1 = require("../../../codegen/oauth/index");
const rxjsStub_1 = require("../../../codegen/oauth/rxjsStub");
const ApiClientConfigurator_1 = require("../../configuration/ApiClientConfigurator");
class OauthDiscovery {
    constructor(config = {}) {
        const configuration = configuration_1.createConfiguration(ApiClientConfigurator_1.ApiClientConfigurator.getParams(config, index_1.ServerConfiguration, rxjsStub_1.Observable, rxjsStub_1.Observable));
        this.accessTokensApi = new index_1.AccessTokensApi(configuration);
        this.refreshTokensApi = new index_1.RefreshTokensApi(configuration);
        this.tokensApi = new index_1.TokensApi(configuration);
    }
    getAuthorizationUrl(clientId, redirectUri, scope, optionalScope, state) {
        const params = {
            client_id: clientId,
            redirect_uri: redirectUri,
            scope,
            optional_scope: optionalScope,
            state,
        };
        return `https://app.hubspot.com/oauth/authorize?${qs.stringify(_.omitBy(params, _.isNil))}`;
    }
}
exports.OauthDiscovery = OauthDiscovery;
