// @prisma/debug/dist/esm/node.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
import tty from "tty";
import util from "util";
exports.init = init;
exports.log = log;
exports.formatArgs = formatArgs;
exports.save = save;
exports.load = load;
exports.useColors = useColors;
exports.destroy = util.deprecate(() => {
}, "Instance method `debug.destroy()` is deprecated and no longer does anything. It will be removed in the next major version of `debug`.");
exports.colors = [6, 2, 3, 4, 5, 1];
exports.inspectOpts = Object.keys(process.env).filter((key) => {
  return /^debug_/i.test(key);
}).reduce((obj, key) => {
  const prop = key.substring(6).toLowerCase().replace(/_([a-z])/g, (_, k) => {
    return k.toUpperCase();
  });
  let val = process.env[key];
  if (/^(yes|on|true|enabled)$/i.test(val)) {
    val = true;
  } else if (/^(no|off|false|disabled)$/i.test(val)) {
    val = false;
  } else if (val === "null") {
    val = null;
  } else {
    val = Number(val);
  }
  obj[prop] = val;
  return obj;
}, {});
function useColors() {
  return "colors" in exports.inspectOpts ? Boolean(exports.inspectOpts.colors) : tty.isatty(process.stderr?.fd);
}
__name(useColors, "useColors");
function formatArgs(args) {
  const { namespace: name, useColors: useColors2 } = this;
  if (useColors2) {
    const c = this.color;
    const colorCode = "[3" + (c < 8 ? c : "8;5;" + c);
    const prefix = `  ${colorCode};1m${name} [0m`;
    args[0] = prefix + args[0].split("\n").join("\n" + prefix);
    args.push(colorCode + "m+" + module.exports.humanize(this.diff) + "[0m");
  } else {
    args[0] = getDate() + name + " " + args[0];
  }
}
__name(formatArgs, "formatArgs");
function getDate() {
  if (exports.inspectOpts.hideDate) {
    return "";
  }
  return new Date().toISOString() + " ";
}
__name(getDate, "getDate");
function log(...args) {
  return process.stderr.write(util.format(...args) + "\n");
}
__name(log, "log");
function save(namespaces) {
  if (namespaces) {
    process.env.DEBUG = namespaces;
  } else {
    delete process.env.DEBUG;
  }
}
__name(save, "save");
function load() {
  return process.env.DEBUG;
}
__name(load, "load");
function init(debug) {
  debug.inspectOpts = {};
  const keys = Object.keys(exports.inspectOpts);
  for (let i = 0; i < keys.length; i++) {
    debug.inspectOpts[keys[i]] = exports.inspectOpts[keys[i]];
  }
}
__name(init, "init");
import { setup } from "./common";
const mod = setup(exports);
module.exports = mod;
var node_default = mod;
const { formatters } = mod;
formatters.o = function(v) {
  this.inspectOpts.colors = this.useColors;
  return util.inspect(v, this.inspectOpts).split("\n").map((str) => str.trim()).join(" ");
};
formatters.O = function(v) {
  this.inspectOpts.colors = this.useColors;
  return util.inspect(v, this.inspectOpts);
};
export {
  node_default as default
};
